<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • hdWGCNA</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><meta property="og:image" content="https://smorabit.github.io/hdWGCNA/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hdWGCNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.04</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/hdWGCNA.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Core functionality</li>
    <li>
      <a href="../articles/basic_tutorial.html">hdWGCNA in single-cell data</a>
    </li>
    <li>
      <a href="../articles/ST_basics.html">hdWGCNA in spatial transcriptomics data</a>
    </li>
    <li>
      <a href="../articles/network_visualizations.html">Network visualization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Biological context for co-expression modules</li>
    <li>
      <a href="../articles/differential_MEs.html">Differential module eigengene (DME) analysis</a>
    </li>
    <li>
      <a href="../articles/module_trait_correlation.html">Module trait correlation</a>
    </li>
    <li>
      <a href="../articles/enrichment_analysis.html">Enrichment analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Exploring modules in external datasets</li>
    <li>
      <a href="../articles/projecting_modules.html">Projecting modules to new datasets</a>
    </li>
    <li>
      <a href="../articles/module_preservation.html">Module preservation and reproducibility</a>
    </li>
    <li>
      <a href="../articles/projecting_modules_cross.html">Cross-species and cross-modality analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced topics</li>
    <li>
      <a href="../articles/consensus_wgcna.html">Consensus network analysis</a>
    </li>
    <li>
      <a href="../articles/motif_analysis.html">Motif analysis</a>
    </li>
    <li>
      <a href="../articles/other_metacells.html">Alternative metacell algorithms</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/customization.html">Module customization</a>
    </li>
    <li>
      <a href="../articles/sctransform.html">Using SCTransform normalized data</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/index.html">All vignettes</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.03" id="hdwgcna-0203-2022-12-15">hdWGCNA 0.2.03 (2022-12-15)<a class="anchor" aria-label="anchor" href="#hdwgcna-0203-2022-12-15"></a></h2>
<div class="section level3">
<h3 id="added-0-2-03">Added<a class="anchor" aria-label="anchor" href="#added-0-2-03"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-2-03">Changes<a class="anchor" aria-label="anchor" href="#changes-0-2-03"></a></h3>
<ul><li>Fixed a bug in <code>ResetModuleNames</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.03" id="hdwgcna-0203-2022-11-10">hdWGCNA 0.2.03 (2022-11-10)<a class="anchor" aria-label="anchor" href="#hdwgcna-0203-2022-11-10"></a></h2>
<div class="section level3">
<h3 id="added-0-2-03-1">Added<a class="anchor" aria-label="anchor" href="#added-0-2-03-1"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-2-03-1">Changes<a class="anchor" aria-label="anchor" href="#changes-0-2-03-1"></a></h3>
<ul><li>New error checking in <code>MetaspotsByGroups</code>
</li>
<li>
<code>MetacellsByGroups</code> now keeps track of which cells were merged.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.02" id="hdwgcna-0202-2022-11-01">hdWGCNA 0.2.02 (2022-11-01)<a class="anchor" aria-label="anchor" href="#hdwgcna-0202-2022-11-01"></a></h2>
<div class="section level3">
<h3 id="added-0-2-02">Added<a class="anchor" aria-label="anchor" href="#added-0-2-02"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-2-02">Changes<a class="anchor" aria-label="anchor" href="#changes-0-2-02"></a></h3>
<ul><li>New warning in <code>SetupForWGCNA</code> if the user selects a very small number of genes.</li>
<li>
<code>MetaspotsByGroups</code> uses sparse matrix format internally.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.01" id="hdwgcna-0201-2022-10-06">hdWGCNA 0.2.01 (2022-10-06)<a class="anchor" aria-label="anchor" href="#hdwgcna-0201-2022-10-06"></a></h2>
<div class="section level3">
<h3 id="added-0-2-01">Added<a class="anchor" aria-label="anchor" href="#added-0-2-01"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-2-01">Changes<a class="anchor" aria-label="anchor" href="#changes-0-2-01"></a></h3>
<ul><li>Bugfix in <code>MetaspotsByGroups</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.00" id="hdwgcna-0200-2022-09-23">hdWGCNA 0.2.00 (2022-09-23)<a class="anchor" aria-label="anchor" href="#hdwgcna-0200-2022-09-23"></a></h2>
<div class="section level3">
<h3 id="added-0-2-00">Added<a class="anchor" aria-label="anchor" href="#added-0-2-00"></a></h3>
<ul><li>
<code>MetaspotsByGroups</code> to aggregate neighboring ST spots prior to network analysis.</li>
<li>Tutorial for applying hdWGCNA to spatial transcriptomics datasets.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-2-00">Changes<a class="anchor" aria-label="anchor" href="#changes-0-2-00"></a></h3>
<ul><li>None</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.2.0001" id="hdwgcna-0120001-2022-09-19">hdWGCNA 0.1.2.0001 (2022-09-19)<a class="anchor" aria-label="anchor" href="#hdwgcna-0120001-2022-09-19"></a></h2>
<div class="section level3">
<h3 id="added-0-1-2-0001">Added<a class="anchor" aria-label="anchor" href="#added-0-1-2-0001"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-2-0001">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-2-0001"></a></h3>
<ul><li>networkType option in <code>TestSoftPowers</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.2.0000" id="hdwgcna-0120000-2022-09-08">hdWGCNA 0.1.2.0000 (2022-09-08)<a class="anchor" aria-label="anchor" href="#hdwgcna-0120000-2022-09-08"></a></h2>
<div class="section level3">
<h3 id="added-0-1-2-0000">Added<a class="anchor" aria-label="anchor" href="#added-0-1-2-0000"></a></h3>
<ul><li>Differential Module Eigengene (DME) tutorial</li>
<li>FindDMEs function</li>
<li>FindAllDMEs function</li>
<li>PlotDMEsVolcano function</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-2-0000">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-2-0000"></a></h3>
<ul><li>None</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9015" id="hdwgcna-0119015-2022-09-06">hdWGCNA 0.1.1.9015 (2022-09-06)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119015-2022-09-06"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9015">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9015"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9015">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9015"></a></h3>
<ul><li>New data format check in <code>MetacellsByGroups</code> to ensure that the selected slot is present in the selected assay.</li>
<li>
<code>SetDatExpr</code> now backs up to the full dataset if the metacell dataset isn’t found.</li>
<li>Changed some text to clarify some points in the Module Preservation tutorial.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9014" id="hdwgcna-0119014-2022-08-26">hdWGCNA 0.1.1.9014 (2022-08-26)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119014-2022-08-26"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9014">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9014"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9014">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9014"></a></h3>
<ul><li>Bugfix in <code>ModuleConnectivity</code> that caused kMEs to be out of order.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9013" id="hdwgcna-0119013-2022-08-25">hdWGCNA 0.1.1.9013 (2022-08-25)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119013-2022-08-25"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9013">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9013"></a></h3>
<ul><li>Tutorial for using SCTransform normalized data.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9013">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9013"></a></h3>
<ul><li>None</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9012" id="hdwgcna-0119012-2022-08-24">hdWGCNA 0.1.1.9012 (2022-08-24)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119012-2022-08-24"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9012">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9012"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9012">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9012"></a></h3>
<ul><li>Now includes options for other types of correlations in <code>ModuleConnectivity</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9011" id="hdwgcna-0119011-2022-08-17">hdWGCNA 0.1.1.9011 (2022-08-17)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119011-2022-08-17"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9011">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9011"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9011">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9011"></a></h3>
<ul><li>Reverted the <code>exclude_grey</code> flag back to doing nothing in the <code>ModuleEigengenes</code> function because it messed up some downstream tasks, will resolve in a future update.</li>
<li>
<code>ProjectModules</code> now excludes modules with too many missing genes in the query dataset, tunable by the <code>overlap_proportion</code> parameter.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9010" id="hdwgcna-0119010-2022-07-30">hdWGCNA 0.1.1.9010 (2022-07-30)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119010-2022-07-30"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9010">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9010"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9010">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9010"></a></h3>
<ul><li>By default, <code>ModuleEigengenes</code> does not compute MEs for the grey module. User can change behavior with the <code>exclude_grey</code> flag.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9009" id="hdwgcna-0119009-2022-07-23">hdWGCNA 0.1.1.9009 (2022-07-23)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119009-2022-07-23"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9009">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9009"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9009">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9009"></a></h3>
<ul><li>Bugfix in <code>ProjectModules</code>.</li>
<li>Bugfix in <code>MetacellsByGroups</code>
</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9008" id="hdwgcna-0119008-2022-07-21">hdWGCNA 0.1.1.9008 (2022-07-21)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119008-2022-07-21"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9008">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9008"></a></h3>
<ul><li>New tutorial for consensus co-expression network analysis.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9008">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9008"></a></h3>
<ul><li>
<code>ModuleNetworkPlot</code> and <code>RunModuleUMAP</code> now checks if <code>ModuleConnectivity</code> has been computed in order to throw a more informative error.</li>
<li>
<code>GetTOM</code> checks if the TOM file exists in order to throw a more informative error.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9007" id="hdwgcna-0119007-2022-07-20">hdWGCNA 0.1.1.9007 (2022-07-20)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119007-2022-07-20"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9007">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9007"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9007">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9007"></a></h3>
<ul><li>New warning message in <code>MetacellsByGroups</code> if there are any groups that are excluded by <code>min_cells</code>.</li>
<li>Assay in Metacell seurat object is now the same as the assay supplied to <code>MetacellsByGroups</code>, instead of the default “RNA”.</li>
<li>
<code>ModuleEigengenes</code> takes <code>assay</code> as an argument, clears up some issues with <code>RunHarmony</code>.</li>
<li>
<code>ModuleEigengenes</code> doesn’t require a “counts” slot to be present in the given assay, but now it throws an error if the normalized data slot is missing.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9006" id="hdwgcna-0119006-2022-07-14">hdWGCNA 0.1.1.9006 (2022-07-14)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119006-2022-07-14"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9006">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9006"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9006">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9006"></a></h3>
<ul><li>
<code>ConstructNetwork</code> now checks if the TOM file already exists, and whether the user wants to overwrite the existing TOM.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9005" id="hdwgcna-0119005-2022-06-17">hdWGCNA 0.1.1.9005 (2022-06-17)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119005-2022-06-17"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9005">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9005"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9005">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9005"></a></h3>
<ul><li>Added new arguments to <code>MetacellsByGroups</code> and <code>ConstructMetacells</code> to exclude very small groups (<code>min_cells</code>), to reach a target number of metacells (<code>target_metacells</code>), and to exclude metacells with too much overlap (<code>max_shared</code>).</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9004" id="hdwgcna-0119004-2022-6-13">hdWGCNA 0.1.1.9004 (2022-6-13)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119004-2022-6-13"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9004">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9004"></a></h3>
<ul><li>None</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9004">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9004"></a></h3>
<ul><li>Bug fix so <code>ModuleTraitCorrelation</code> can be run with a single trait.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9003" id="hdwgcna-0119003-2022-6-11">hdWGCNA 0.1.1.9003 (2022-6-11)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119003-2022-6-11"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9003">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9003"></a></h3>
<ul><li>
<code>GetHubGenes</code> function to extract the top hub genes from the module assignment table.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9003">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9003"></a></h3>
<ul><li>
<code>ConstructMetacells</code> stores run statistics as a table, and has a new option to exclude metacells that overlap with each other.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9002" id="hdwgcna-0119002-2022-5-24">hdWGCNA 0.1.1.9002 (2022-5-24)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119002-2022-5-24"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9002">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9002"></a></h3>
<ul><li>
<code>ModuleEigengenes</code> checks to make sure the data has been scaled.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9002">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9002"></a></h3>
<ul><li>None</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9001" id="hdwgcna-0119001-2022-05-09">hdWGCNA 0.1.1.9001 (2022-05-09)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119001-2022-05-09"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9001">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9001"></a></h3>
<ul><li>Wrote docstring for <code>PlotKMEs</code> so it would actually be included in the package.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9001">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9001"></a></h3>
<ul><li>None</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1.9000" id="hdwgcna-0119000-2022-05-05">hdWGCNA 0.1.1.9000 (2022-05-05)<a class="anchor" aria-label="anchor" href="#hdwgcna-0119000-2022-05-05"></a></h2>
<div class="section level3">
<h3 id="added-0-1-1-9000">Added<a class="anchor" aria-label="anchor" href="#added-0-1-1-9000"></a></h3>
<ul><li>
<code>PlotKMEs</code> function to visualize the genes in a module ranked by kME.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-0-1-1-9000">Changes<a class="anchor" aria-label="anchor" href="#changes-0-1-1-9000"></a></h3>
<ul><li>Changed the name of the package from scWGCNA to hdWGCNA, since we plan to accommodate data types beyond single cell alone.</li>
<li>Updated tutorial to recommend computing kMEs in a specific cell population.</li>
</ul></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

