<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Vignettes overview â€¢ hdWGCNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Vignettes overview">
<meta property="og:description" content="Document showing an overview of the different analysis vignettes.
">
<meta property="og:image" content="https://smorabit.github.io/hdWGCNA/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hdWGCNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.24</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/hdWGCNA.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Core functionality</li>
    <li>
      <a href="../articles/basic_tutorial.html">hdWGCNA in single-cell data</a>
    </li>
    <li>
      <a href="../articles/ST_basics.html">hdWGCNA in spatial transcriptomics data</a>
    </li>
    <li>
      <a href="../articles/isoform_pbmcs.html">Isoform co-expression network analysis with PacBio MAS-Seq</a>
    </li>
    <li>
      <a href="../articles/network_visualizations.html">Network visualization</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Biological context for co-expression modules</li>
    <li>
      <a href="../articles/differential_MEs.html">Differential module eigengene (DME) analysis</a>
    </li>
    <li>
      <a href="../articles/module_trait_correlation.html">Module trait correlation</a>
    </li>
    <li>
      <a href="../articles/enrichment_analysis.html">Enrichment analysis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Exploring modules in external datasets</li>
    <li>
      <a href="../articles/projecting_modules.html">Projecting modules to new datasets</a>
    </li>
    <li>
      <a href="../articles/module_preservation.html">Module preservation and reproducibility</a>
    </li>
    <li>
      <a href="../articles/projecting_modules_cross.html">Cross-species and cross-modality analysis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Advanced topics</li>
    <li>
      <a href="../articles/consensus_wgcna.html">Consensus network analysis</a>
    </li>
    <li>
      <a href="../articles/pseudobulk.html">hdWGCNA with pseudobulk data</a>
    </li>
    <li>
      <a href="../articles/pseudotime.html">Co-expression module dynamics with pseudotime</a>
    </li>
    <li>
      <a href="../articles/motif_analysis.html">Motif analysis</a>
    </li>
    <li>
      <a href="../articles/other_metacells.html">Alternative metacell algorithms</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/customization.html">Module customization</a>
    </li>
    <li>
      <a href="../articles/sctransform.html">Using SCTransform normalized data</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/index.html">All vignettes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Vignettes overview</h1>
            
      
      
      <div class="hidden name"><code>hdWGCNA.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="co-expression-network-analysis">Co-expression network analysis<a class="anchor" aria-label="anchor" href="#co-expression-network-analysis"></a>
</h2>
<p>These tutorials cover the essentials of performing co-expression network analysis in single-cell transcriptomics data, and visualizing the key results.</p>
<div class="section level3">
<h3 id="hdwgcna-in-single-cell-data">
<a href="basic_tutorial.html">hdWGCNA in single-cell data</a><a class="anchor" aria-label="anchor" href="#hdwgcna-in-single-cell-data"></a>
</h3>
<p>This tutorial covers the essential functions to construct a co-expression network in single-cell transcriptomics data with hdWGCNA.</p>
<p><img src="figures/basic_tutorial/Zhou_featureplot_hMEs_selected_wide.png" width="500" height="500"></p>
</div>
<div class="section level3">
<h3 id="hdwgcna-in-spatial-transcriptomics-data">
<a href="ST_basics.html">hdWGCNA in spatial transcriptomics data</a><a class="anchor" aria-label="anchor" href="#hdwgcna-in-spatial-transcriptomics-data"></a>
</h3>
<p>This tutorial covers the essential functions to construct a co-expression network in spatial transcriptomics data with hdWGCNA.</p>
<p><img src="figures/ST_basics/spatial_clusters.png" width="500" height="500"></p>
</div>
<div class="section level3">
<h3 id="network-visualization">
<a href="network_visualizations.html">Network visualization</a><a class="anchor" aria-label="anchor" href="#network-visualization"></a>
</h3>
<p>This tutorial highlights several approaches for visualizing the hdWGCNA co-expression networks.</p>
<p><img src="figures/network_vis/hubgene_umap_igraph.png" width="400" height="400"></p>
</div>
</div>
<div class="section level2">
<h2 id="biological-context-for-co-expression-modules">Biological context for co-expression modules<a class="anchor" aria-label="anchor" href="#biological-context-for-co-expression-modules"></a>
</h2>
<p>These tutorials will provide further biological context for our co-expression modules, potentially revealing what experimental conditions and biological processes that these modules are involved in.</p>
<div class="section level3">
<h3 id="differential-module-eigengene-dme-analysis">
<a href="differential_MEs.html">Differential module eigengene (DME) analysis</a><a class="anchor" aria-label="anchor" href="#differential-module-eigengene-dme-analysis"></a>
</h3>
<p>This tutorial covers how to compare module eigengenes between experimental groups.</p>
<p><img src="figures/DMEs/test_DME_volcano.png" width="400" height="400" href="differential_MEs.html"></p>
</div>
<div class="section level3">
<h3 id="module-trait-correlation">
<a href="module_trait_correlation.html">Module trait correlation</a><a class="anchor" aria-label="anchor" href="#module-trait-correlation"></a>
</h3>
<p>This tutorial covers how to correlate continuous and categorical variables with module eigengenes or module expression scores, revealing which modules are related to different experimental conditions or covariates.</p>
<p><img src="figures/mt_correlation/ME_Trait_correlation_fdr.png" width="400" height="400" href="module_trait_correlation.html"></p>
</div>
<div class="section level3">
<h3 id="enrichment-analysis">
<a href="enrichment_analysis.html">Enrichment analysis</a><a class="anchor" aria-label="anchor" href="#enrichment-analysis"></a>
</h3>
<p>This tutorial shows how to use Enrichr to compare the gene members of each co-expression module to curated gene lists, thereby pointing towards the biological functions of the co-expression modules.</p>
<p><img src="figures/enrichment/GO_dotplot.png" width="400" height="400"></p>
</div>
</div>
<div class="section level2">
<h2 id="exploring-modules-in-external-datasets">Exploring modules in external datasets<a class="anchor" aria-label="anchor" href="#exploring-modules-in-external-datasets"></a>
</h2>
<div class="section level3">
<h3 id="projecting-modules-to-new-datasets">
<a href="projecting_modules.html">Projecting modules to new datasets</a><a class="anchor" aria-label="anchor" href="#projecting-modules-to-new-datasets"></a>
</h3>
<p>This tutorial covers how to project co-expression modules from a reference to a query dataset.</p>
<p><img src="figures/projection/compare_umaps.png" width="500" height="500"></p>
</div>
<div class="section level3">
<h3 id="module-preservation-and-reproducibility">
<a href="module_preservation.html">Module preservation and reproducibility</a><a class="anchor" aria-label="anchor" href="#module-preservation-and-reproducibility"></a>
</h3>
<p>This tutorial covers statistical methods for assessing the preservation and reproducibility of co-expression networks using external datasets.</p>
<p><img src="figures/projection/module_preservation_summary.png" width="500" height="500"></p>
</div>
<div class="section level3">
<h3 id="cross-species-and-cross-modality-analysis">
<a href="projecting_modules_cross.html">Cross-species and cross-modality analysis</a><a class="anchor" aria-label="anchor" href="#cross-species-and-cross-modality-analysis"></a>
</h3>
<p>This tutorial covers how to project co-expression modules from a reference to a query dataset for special cases where the data modality or the species do not match between the reference and the query.</p>
<p><img src="figures/projection/atac_umap_covplot.png" width="700" height="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="advanced-topics">Advanced topics<a class="anchor" aria-label="anchor" href="#advanced-topics"></a>
</h2>
<div class="section level3">
<h3 id="consensus-network-analysis">
<a href="consensus_wgcna.html">Consensus network analysis</a><a class="anchor" aria-label="anchor" href="#consensus-network-analysis"></a>
</h3>
</div>
<div class="section level3">
<h3 id="motif-analysis">
<a href="motif_analysis.html">Motif analysis</a><a class="anchor" aria-label="anchor" href="#motif-analysis"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="other">Other<a class="anchor" aria-label="anchor" href="#other"></a>
</h2>
<div class="section level3">
<h3 id="module-customization">
<a href="customization.html">Module customization</a><a class="anchor" aria-label="anchor" href="#module-customization"></a>
</h3>
<p>This tutorial covers how to change the default names and colors for hdWGCNA modules.</p>
<p><img src="figures/customization/featureplot.png" width="400" height="400"></p>
</div>
<div class="section level3">
<h3 id="using-sctransform-normalized-data">
<a href="sctransform.html">Using SCTransform normalized data</a><a class="anchor" aria-label="anchor" href="#using-sctransform-normalized-data"></a>
</h3>
<p>This tutorial covers how to use SCTransform normalized data in hdWGCNA.</p>
<p><img src="figures/sctransform/dendro_SCT_compare.png" width="400" height="400" href="sctransform.html"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
