<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Alternative metacell algorithms • hdWGCNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Alternative metacell algorithms">
<meta property="og:description" content="Tutorial for performing co-expression network analysis using alternative metacell aggregation algorithms.
">
<meta property="og:image" content="https://smorabit.github.io/hdWGCNA/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hdWGCNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1.9013</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/hdWGCNA.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Core functionality</li>
    <li>
      <a href="../articles/basic_tutorial.html">hdWGCNA basics</a>
    </li>
    <li>
      <a href="../articles/network_visualizations.html">Network visualization</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Biological context for co-expression modules</li>
    <li>
      <a href="../articles/module_trait_correlation.html">Module trait correlation</a>
    </li>
    <li>
      <a href="../articles/enrichment_analysis.html">Enrichment analysis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Exploring modules in external datasets</li>
    <li>
      <a href="../articles/projecting_modules.html">Projecting modules to new datasets</a>
    </li>
    <li>
      <a href="../articles/module_preservation.html">Module preservation and reproducibility</a>
    </li>
    <li>
      <a href="../articles/projecting_modules_cross.html">Cross-species and cross-modality analysis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Advanced topics</li>
    <li>
      <a href="../articles/consensus_wgcna.html">Consensus network analysis</a>
    </li>
    <li>
      <a href="../articles/motif_analysis.html">Motif analysis</a>
    </li>
    <li>
      <a href="../articles/other_metacells.html">Alternative metacell algorithms</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/customization.html">Module customization</a>
    </li>
    <li>
      <a href="../articles/sctransform.html">Using SCTransform normalized data</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/index.html">All vignettes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Alternative metacell algorithms</h1>
            
      
      
      <div class="hidden name"><code>other_metacells.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we demonstrate an advanced analysis showing how to use alternative metacell aggregation algorithms for co-expression network analysis. Here we will identify metacells using <a href="https://github.com/dpeerlab/SEACells" class="external-link">SEACells</a> and <a href="https://github.com/tanaylab/metacells" class="external-link">Metacell2</a> (MC2) using the CD34+ hematopoietic stem and progenitor stem cells dataset provided with SEACells. Both SEACells and MC2 are Python packages, so they are not directly interoperable with hdWGCNA and are not formally included as part of the hdWGCNA package. Thus, in this tutorial we will follow the recommended workflows for SEACells and MC2 in Python, then export the resulting data so they can be loaded into R.</p>
<p>Before getting started with this tutorial, please install SEACells and MC2 using the github links above. I was able to create a conda environment for SEACells as specified in their github, and then install MC2 within that environment.</p>
<p>We can download the practice dataset provided with SEACells for this tutorial:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://dp-lab-data-public.s3.amazonaws.com/SEACells-multiome/cd34_multiome_rna.h5ad -O cd34_multiome_rna.h5ad</span></code></pre></div>
<div class="section level2">
<h2 id="constructing-metacells-with-seacells">Constructing metacells with SEACells<a class="anchor" aria-label="anchor" href="#constructing-metacells-with-seacells"></a>
</h2>
<p>In this section, we follow the recommended workflow for constructing metacells with SEACells. Here we only show the code, but you may wish to consult the <a href="https://github.com/dpeerlab/SEACells/blob/main/notebooks/SEACell_computation.ipynb" class="external-link">tutorial from the SEACells github page</a> for more information.</p>
<p><strong><em>Note</em></strong> I tried to run this tutorial on a larger dataset of 60k cells and 30k genes, but SEACells didn’t complete running after several hours. Thus, your mileage may vary depending on your input dataset, and that issue is essentially the reason why this entire tutorial is done using the SEACells hematopoietic stem cell dataset.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># following this tutorial: https://github.com/dpeerlab/SEACells/blob/main/notebooks/SEACell_computation.ipynb</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> SEACells</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy <span class="im">import</span> io</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="co"># load the dataset downloaded above into Python</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">'cd34_multiome_rna.h5ad'</span>)</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="co"># retain the unprocessed UMI counts matrix in the .raw slot</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>raw_ad <span class="op">=</span> sc.AnnData(adata.X)</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>raw_ad.obs_names, raw_ad.var_names <span class="op">=</span> adata.obs_names, adata.var_names</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>adata.raw <span class="op">=</span> raw_ad</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="co"># process data with SCANPY</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a><span class="co"># note that we don't scale the data matrix before PCA. this is how</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a><span class="co"># they do it in the SEACells tutorial so we do it that way here.</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_per_cell(adata)</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata)</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(adata, n_top_genes<span class="op">=</span><span class="dv">1500</span>)</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(adata, n_comps<span class="op">=</span><span class="dv">50</span>, use_highly_variable<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a><span class="co">##################################################################################</span></span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Running SEACells</span></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a><span class="co">##################################################################################</span></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a><span class="co"># they recommend one metacell for every 75 real cells</span></span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>n_SEACells <span class="op">=</span> <span class="bu">int</span>(np.floor(adata.obs.shape[<span class="dv">0</span>] <span class="op">/</span> <span class="dv">75</span>))</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>build_kernel_on <span class="op">=</span> <span class="st">'X_pca'</span> <span class="co"># key in ad.obsm to use for computing metacells</span></span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>                          <span class="co"># This would be replaced by 'X_svd' for ATAC data</span></span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a><span class="co">## Additional parameters</span></span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a>n_waypoint_eigs <span class="op">=</span> <span class="dv">10</span> <span class="co"># Number of eigenvalues to consider when initializing metacells</span></span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a>waypoint_proportion <span class="op">=</span> <span class="fl">0.9</span> <span class="co"># Proportion of metacells to initialize using waypoint analysis,</span></span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a>                        <span class="co"># the remainder of cells are selected by greedy selection</span></span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a><span class="co"># set up the model</span></span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> SEACells.core.SEACells(adata,</span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a>                  build_kernel_on<span class="op">=</span>build_kernel_on,</span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a>                  n_SEACells<span class="op">=</span>n_SEACells,</span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a>                  n_waypoint_eigs<span class="op">=</span>n_waypoint_eigs,</span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a>                  waypt_proportion<span class="op">=</span>waypoint_proportion,</span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a>                  convergence_epsilon <span class="op">=</span> <span class="fl">1e-5</span>)</span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-53"><a href="#cb2-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize archetypes</span></span>
<span id="cb2-54"><a href="#cb2-54" aria-hidden="true" tabindex="-1"></a>model.initialize_archetypes()</span>
<span id="cb2-55"><a href="#cb2-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-56"><a href="#cb2-56" aria-hidden="true" tabindex="-1"></a><span class="co"># fit the model</span></span>
<span id="cb2-57"><a href="#cb2-57" aria-hidden="true" tabindex="-1"></a>model.fit(n_iter<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb2-58"><a href="#cb2-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-59"><a href="#cb2-59" aria-hidden="true" tabindex="-1"></a><span class="co"># create aggregated metacell expression dataset</span></span>
<span id="cb2-60"><a href="#cb2-60" aria-hidden="true" tabindex="-1"></a>SEACell_ad <span class="op">=</span> SEACells.core.summarize_by_SEACell(adata, SEACells_label<span class="op">=</span><span class="st">'SEACell'</span>, summarize_layer<span class="op">=</span><span class="st">'raw'</span>)</span></code></pre></div>
<p>Next, we need to write the data matrices to disk so we can read them into R. There are several different approaches to convert directly between <code>.h5ad</code> and <code>Seurat</code> formats, but I have personally run into a lot of unresolved bugs with these approaches (such as <a href="https://github.com/mojaveazure/seurat-disk" class="external-link">SeuratDisk</a>) so exporting the data using <code>scipy</code> and <code>pandas</code> is more reliable, at least in my experience.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># write the h5ad files in case we want to load them into Python again</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(<span class="st">'data/tutorial_singlecell.h5ad'</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>SEACell_ad.write_h5ad(<span class="st">'data/tutorial_seacells.h5ad'</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Save components of single-cell dataset</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># write obs table</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>data_dir <span class="op">=</span> <span class="st">'data/'</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'UMAP_1'</span>] <span class="op">=</span> adata.obsm[<span class="st">'X_umap'</span>][:,<span class="dv">0</span>]</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'UMAP_2'</span>] <span class="op">=</span> adata.obsm[<span class="st">'X_umap'</span>][:,<span class="dv">1</span>]</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>adata.obs.to_csv(<span class="st">'</span><span class="sc">{}</span><span class="st">/tutorial_singlecell_obs.csv'</span>.<span class="bu">format</span>(data_dir))</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="co"># write var table:</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>adata.var.to_csv(<span class="st">'</span><span class="sc">{}</span><span class="st">/tutorial_singlecell_var.csv'</span>.<span class="bu">format</span>(data_dir))</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="co"># save the sparse matrix for Seurat:</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>X <span class="op">=</span> adata.raw.X</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>X <span class="op">=</span> scipy.sparse.csr_matrix.transpose(X)</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>io.mmwrite(<span class="st">'</span><span class="sc">{}</span><span class="st">/tutorial_singlecell.mtx'</span>.<span class="bu">format</span>(data_dir), X)</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="co"># write the PCA</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>pd.DataFrame(adata.obsm[<span class="st">'X_pca'</span>]).to_csv(<span class="st">'</span><span class="sc">{}</span><span class="st">/tutorial_singlecell_pca.csv'</span>.<span class="bu">format</span>(data_dir))</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Save components of SEACells dataset</span></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a><span class="co"># write obs table</span></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>SEACell_ad.obs.to_csv(<span class="st">'</span><span class="sc">{}</span><span class="st">/tutorial_seacells_obs.csv'</span>.<span class="bu">format</span>(data_dir))</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a><span class="co"># save the sparse matrix for Seurat:</span></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>X <span class="op">=</span> SEACell_ad.X</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>X <span class="op">=</span> scipy.sparse.csr_matrix.transpose(X)</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>io.mmwrite(<span class="st">'</span><span class="sc">{}</span><span class="st">/tutorial_seacells.mtx'</span>.<span class="bu">format</span>(data_dir), X)</span></code></pre></div>
<p>If you just want to run SEACells and not MC2, you can skip the next section.</p>
</div>
<div class="section level2">
<h2 id="constructing-metacells-with-mc2">Constructing metacells with MC2<a class="anchor" aria-label="anchor" href="#constructing-metacells-with-mc2"></a>
</h2>
<p>In this section, we follow the <a href="https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html" class="external-link">recommended workflow for constructing metacells with MC2</a>, using the same dataset that was used for SEACells.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># following this tutorial: https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> SEACells</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy <span class="im">import</span> io</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> metacells <span class="im">as</span> mc</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.sparse <span class="im">as</span> sp</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sb</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> math <span class="im">import</span> hypot</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">'data/cd34_multiome_rna.h5ad'</span>)</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="co"># need to run these utilities functions to fix the counts matrix</span></span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>X <span class="op">=</span> adata.X</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>mc.utilities.typing.sum_duplicates(X)</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>mc.utilities.typing.sort_indices(X)</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>adata.X <span class="op">=</span> X</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a><span class="co"># set the raw counts matrix</span></span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a>raw_ad <span class="op">=</span> sc.AnnData(adata.X)</span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>raw_ad.obs_names, raw_ad.var_names <span class="op">=</span> adata.obs_names, adata.var_names</span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>adata.raw <span class="op">=</span> raw_ad</span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a><span class="co"># name of the dataset</span></span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a>mc.ut.set_name(adata, <span class="st">'cd34'</span>)</span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Cleaning genes</span></span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a>excluded_gene_names <span class="op">=</span> [<span class="st">'IGHMBP2'</span>, <span class="st">'IGLL1'</span>, <span class="st">'IGLL5'</span>, <span class="st">'IGLON5'</span>, <span class="st">'NEAT1'</span>, <span class="st">'TMSB10'</span>, <span class="st">'TMSB4X'</span>]</span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>excluded_gene_patterns <span class="op">=</span> [<span class="st">'MT-.*'</span>]</span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>mc.pl.analyze_clean_genes(adata,</span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_names<span class="op">=</span>excluded_gene_names,</span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_patterns<span class="op">=</span>excluded_gene_patterns,</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a>                          random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a><span class="co"># combine into a clean gene mask</span></span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a>mc.pl.pick_clean_genes(adata)</span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a><span class="co"># Clean cells</span></span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a>full <span class="op">=</span> adata</span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a>properly_sampled_min_cell_total <span class="op">=</span> <span class="dv">800</span></span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_cell_total <span class="op">=</span> <span class="dv">15000</span></span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a>total_umis_of_cells <span class="op">=</span> mc.ut.get_o_numpy(full, name<span class="op">=</span><span class="st">'__x__'</span>, <span class="bu">sum</span><span class="op">=</span><span class="va">True</span>)</span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>too_small_cells_count <span class="op">=</span> <span class="bu">sum</span>(total_umis_of_cells <span class="op">&lt;</span> properly_sampled_min_cell_total)</span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a>too_large_cells_count <span class="op">=</span> <span class="bu">sum</span>(total_umis_of_cells <span class="op">&gt;</span> properly_sampled_max_cell_total)</span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a>too_small_cells_percent <span class="op">=</span> <span class="fl">100.0</span> <span class="op">*</span> too_small_cells_count <span class="op">/</span> <span class="bu">len</span>(total_umis_of_cells)</span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a>too_large_cells_percent <span class="op">=</span> <span class="fl">100.0</span> <span class="op">*</span> too_large_cells_count <span class="op">/</span> <span class="bu">len</span>(total_umis_of_cells)</span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Will exclude %s (%.2f%%) cells with less than %s UMIs"</span></span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a>      <span class="op">%</span> (too_small_cells_count,</span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a>         too_small_cells_percent,</span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a>         properly_sampled_min_cell_total))</span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Will exclude %s (%.2f%%) cells with more than %s UMIs"</span></span>
<span id="cb4-80"><a href="#cb4-80" aria-hidden="true" tabindex="-1"></a>      <span class="op">%</span> (too_large_cells_count,</span>
<span id="cb4-81"><a href="#cb4-81" aria-hidden="true" tabindex="-1"></a>         too_large_cells_percent,</span>
<span id="cb4-82"><a href="#cb4-82" aria-hidden="true" tabindex="-1"></a>         properly_sampled_max_cell_total))</span>
<span id="cb4-83"><a href="#cb4-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-84"><a href="#cb4-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-85"><a href="#cb4-85" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_excluded_genes_fraction <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb4-86"><a href="#cb4-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-87"><a href="#cb4-87" aria-hidden="true" tabindex="-1"></a>excluded_genes_data <span class="op">=</span> mc.tl.filter_data(full, var_masks<span class="op">=</span>[<span class="st">'~clean_gene'</span>])[<span class="dv">0</span>]</span>
<span id="cb4-88"><a href="#cb4-88" aria-hidden="true" tabindex="-1"></a>excluded_umis_of_cells <span class="op">=</span> mc.ut.get_o_numpy(excluded_genes_data, name<span class="op">=</span><span class="st">'__x__'</span>, <span class="bu">sum</span><span class="op">=</span><span class="va">True</span>)</span>
<span id="cb4-89"><a href="#cb4-89" aria-hidden="true" tabindex="-1"></a>excluded_fraction_of_umis_of_cells <span class="op">=</span> excluded_umis_of_cells <span class="op">/</span> total_umis_of_cells</span>
<span id="cb4-90"><a href="#cb4-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-91"><a href="#cb4-91" aria-hidden="true" tabindex="-1"></a>too_excluded_cells_count <span class="op">=</span> <span class="bu">sum</span>(excluded_fraction_of_umis_of_cells <span class="op">&gt;</span> properly_sampled_max_excluded_genes_fraction)</span>
<span id="cb4-92"><a href="#cb4-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-93"><a href="#cb4-93" aria-hidden="true" tabindex="-1"></a>too_excluded_cells_percent <span class="op">=</span> <span class="fl">100.0</span> <span class="op">*</span> too_excluded_cells_count <span class="op">/</span> <span class="bu">len</span>(total_umis_of_cells)</span>
<span id="cb4-94"><a href="#cb4-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-95"><a href="#cb4-95" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Will exclude %s (%.2f%%) cells with more than %.2f%% excluded gene UMIs"</span></span>
<span id="cb4-96"><a href="#cb4-96" aria-hidden="true" tabindex="-1"></a>      <span class="op">%</span> (too_excluded_cells_count,</span>
<span id="cb4-97"><a href="#cb4-97" aria-hidden="true" tabindex="-1"></a>         too_excluded_cells_percent,</span>
<span id="cb4-98"><a href="#cb4-98" aria-hidden="true" tabindex="-1"></a>         <span class="fl">100.0</span> <span class="op">*</span> properly_sampled_max_excluded_genes_fraction))</span>
<span id="cb4-99"><a href="#cb4-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-100"><a href="#cb4-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-101"><a href="#cb4-101" aria-hidden="true" tabindex="-1"></a>mc.pl.analyze_clean_cells(</span>
<span id="cb4-102"><a href="#cb4-102" aria-hidden="true" tabindex="-1"></a>    full,</span>
<span id="cb4-103"><a href="#cb4-103" aria-hidden="true" tabindex="-1"></a>    properly_sampled_min_cell_total<span class="op">=</span>properly_sampled_min_cell_total,</span>
<span id="cb4-104"><a href="#cb4-104" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_cell_total<span class="op">=</span>properly_sampled_max_cell_total,</span>
<span id="cb4-105"><a href="#cb4-105" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_excluded_genes_fraction<span class="op">=</span>properly_sampled_max_excluded_genes_fraction)</span>
<span id="cb4-106"><a href="#cb4-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-107"><a href="#cb4-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-108"><a href="#cb4-108" aria-hidden="true" tabindex="-1"></a>mc.pl.pick_clean_cells(full)</span>
<span id="cb4-109"><a href="#cb4-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-110"><a href="#cb4-110" aria-hidden="true" tabindex="-1"></a>clean <span class="op">=</span> mc.pl.extract_clean_data(full)</span>
<span id="cb4-111"><a href="#cb4-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-112"><a href="#cb4-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-113"><a href="#cb4-113" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb4-114"><a href="#cb4-114" aria-hidden="true" tabindex="-1"></a><span class="co"># Forbidden genes</span></span>
<span id="cb4-115"><a href="#cb4-115" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb4-116"><a href="#cb4-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-117"><a href="#cb4-117" aria-hidden="true" tabindex="-1"></a>suspect_gene_names <span class="op">=</span> [<span class="st">'PCNA'</span>, <span class="st">'MKI67'</span>, <span class="st">'TOP2A'</span>, <span class="st">'HIST1H1D'</span>,</span>
<span id="cb4-118"><a href="#cb4-118" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'FOS'</span>, <span class="st">'JUN'</span>, <span class="st">'HSP90AB1'</span>, <span class="st">'HSPA1A'</span>,</span>
<span id="cb4-119"><a href="#cb4-119" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'ISG15'</span>, <span class="st">'WARS'</span> ]</span>
<span id="cb4-120"><a href="#cb4-120" aria-hidden="true" tabindex="-1"></a>suspect_gene_patterns <span class="op">=</span> [ <span class="st">'MCM[0-9]'</span>, <span class="st">'SMC[0-9]'</span>, <span class="st">'IFI.*'</span> ]</span>
<span id="cb4-121"><a href="#cb4-121" aria-hidden="true" tabindex="-1"></a>suspect_genes_mask <span class="op">=</span> mc.tl.find_named_genes(clean, names<span class="op">=</span>suspect_gene_names,</span>
<span id="cb4-122"><a href="#cb4-122" aria-hidden="true" tabindex="-1"></a>                                            patterns<span class="op">=</span>suspect_gene_patterns)</span>
<span id="cb4-123"><a href="#cb4-123" aria-hidden="true" tabindex="-1"></a>suspect_gene_names <span class="op">=</span> <span class="bu">sorted</span>(clean.var_names[suspect_genes_mask])</span>
<span id="cb4-124"><a href="#cb4-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-125"><a href="#cb4-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-126"><a href="#cb4-126" aria-hidden="true" tabindex="-1"></a>mc.pl.relate_genes(clean, random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb4-127"><a href="#cb4-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-128"><a href="#cb4-128" aria-hidden="true" tabindex="-1"></a><span class="co"># which groups of genes contain sus genes</span></span>
<span id="cb4-129"><a href="#cb4-129" aria-hidden="true" tabindex="-1"></a>module_of_genes <span class="op">=</span> clean.var[<span class="st">'related_genes_module'</span>]</span>
<span id="cb4-130"><a href="#cb4-130" aria-hidden="true" tabindex="-1"></a>suspect_gene_modules <span class="op">=</span> np.unique(module_of_genes[suspect_genes_mask])</span>
<span id="cb4-131"><a href="#cb4-131" aria-hidden="true" tabindex="-1"></a>suspect_gene_modules <span class="op">=</span> suspect_gene_modules[suspect_gene_modules <span class="op">&gt;=</span> <span class="dv">0</span>]</span>
<span id="cb4-132"><a href="#cb4-132" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(suspect_gene_modules)</span>
<span id="cb4-133"><a href="#cb4-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-134"><a href="#cb4-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-135"><a href="#cb4-135" aria-hidden="true" tabindex="-1"></a>similarity_of_genes <span class="op">=</span> mc.ut.get_vv_frame(clean, <span class="st">'related_genes_similarity'</span>)</span>
<span id="cb4-136"><a href="#cb4-136" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> gene_module <span class="kw">in</span> suspect_gene_modules:</span>
<span id="cb4-137"><a href="#cb4-137" aria-hidden="true" tabindex="-1"></a>    module_genes_mask <span class="op">=</span> module_of_genes <span class="op">==</span> gene_module</span>
<span id="cb4-138"><a href="#cb4-138" aria-hidden="true" tabindex="-1"></a>    similarity_of_module <span class="op">=</span> similarity_of_genes.loc[module_genes_mask, module_genes_mask]</span>
<span id="cb4-139"><a href="#cb4-139" aria-hidden="true" tabindex="-1"></a>    similarity_of_module.index <span class="op">=</span> <span class="op">\</span></span>
<span id="cb4-140"><a href="#cb4-140" aria-hidden="true" tabindex="-1"></a>    similarity_of_module.columns <span class="op">=</span> [</span>
<span id="cb4-141"><a href="#cb4-141" aria-hidden="true" tabindex="-1"></a>        <span class="st">'(*) '</span> <span class="op">+</span> name <span class="cf">if</span> name <span class="kw">in</span> suspect_gene_names <span class="cf">else</span> name</span>
<span id="cb4-142"><a href="#cb4-142" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> name <span class="kw">in</span> similarity_of_module.index</span>
<span id="cb4-143"><a href="#cb4-143" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb4-144"><a href="#cb4-144" aria-hidden="true" tabindex="-1"></a>    ax <span class="op">=</span> plt.axes()</span>
<span id="cb4-145"><a href="#cb4-145" aria-hidden="true" tabindex="-1"></a>    sb.heatmap(similarity_of_module, vmin<span class="op">=</span><span class="dv">0</span>, vmax<span class="op">=</span><span class="dv">1</span>, xticklabels<span class="op">=</span><span class="va">True</span>, yticklabels<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>ax, cmap<span class="op">=</span><span class="st">"YlGnBu"</span>)</span>
<span id="cb4-146"><a href="#cb4-146" aria-hidden="true" tabindex="-1"></a>    ax.set_title(<span class="ss">f'Gene Module </span><span class="sc">{</span>gene_module<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb4-147"><a href="#cb4-147" aria-hidden="true" tabindex="-1"></a>    plt.savefig(<span class="st">'figures/module_heatmap_</span><span class="sc">{}</span><span class="st">.pdf'</span>.<span class="bu">format</span>(gene_module))</span>
<span id="cb4-148"><a href="#cb4-148" aria-hidden="true" tabindex="-1"></a>    plt.clf()</span>
<span id="cb4-149"><a href="#cb4-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-150"><a href="#cb4-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-151"><a href="#cb4-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-152"><a href="#cb4-152" aria-hidden="true" tabindex="-1"></a><span class="co"># genes that are correlated with the known forbidden genes</span></span>
<span id="cb4-153"><a href="#cb4-153" aria-hidden="true" tabindex="-1"></a>forbidden_genes_mask <span class="op">=</span> suspect_genes_mask</span>
<span id="cb4-154"><a href="#cb4-154" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> gene_module <span class="kw">in</span> [<span class="dv">17</span>, <span class="dv">19</span>, <span class="dv">24</span>, <span class="dv">78</span>, <span class="dv">113</span>, <span class="dv">144</span>, <span class="dv">149</span>]:</span>
<span id="cb4-155"><a href="#cb4-155" aria-hidden="true" tabindex="-1"></a>    module_genes_mask <span class="op">=</span> module_of_genes <span class="op">==</span> gene_module</span>
<span id="cb4-156"><a href="#cb4-156" aria-hidden="true" tabindex="-1"></a>    forbidden_genes_mask <span class="op">|=</span> module_genes_mask</span>
<span id="cb4-157"><a href="#cb4-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-158"><a href="#cb4-158" aria-hidden="true" tabindex="-1"></a>forbidden_gene_names <span class="op">=</span> <span class="bu">sorted</span>(clean.var_names[forbidden_genes_mask])</span>
<span id="cb4-159"><a href="#cb4-159" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(forbidden_gene_names))</span>
<span id="cb4-160"><a href="#cb4-160" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">' '</span>.join(forbidden_gene_names))</span>
<span id="cb4-161"><a href="#cb4-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-162"><a href="#cb4-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-163"><a href="#cb4-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-164"><a href="#cb4-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-165"><a href="#cb4-165" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb4-166"><a href="#cb4-166" aria-hidden="true" tabindex="-1"></a><span class="co"># computing metacells</span></span>
<span id="cb4-167"><a href="#cb4-167" aria-hidden="true" tabindex="-1"></a><span class="co">################################################################################</span></span>
<span id="cb4-168"><a href="#cb4-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-169"><a href="#cb4-169" aria-hidden="true" tabindex="-1"></a>max_parallel_piles <span class="op">=</span> mc.pl.guess_max_parallel_piles(clean)</span>
<span id="cb4-170"><a href="#cb4-170" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(max_parallel_piles)</span>
<span id="cb4-171"><a href="#cb4-171" aria-hidden="true" tabindex="-1"></a>mc.pl.set_max_parallel_piles(max_parallel_piles)</span>
<span id="cb4-172"><a href="#cb4-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-173"><a href="#cb4-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-174"><a href="#cb4-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-175"><a href="#cb4-175" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mc.ut.progress_bar():</span>
<span id="cb4-176"><a href="#cb4-176" aria-hidden="true" tabindex="-1"></a>    mc.pl.divide_and_conquer_pipeline(clean,</span>
<span id="cb4-177"><a href="#cb4-177" aria-hidden="true" tabindex="-1"></a>                                      forbidden_gene_names<span class="op">=</span>forbidden_gene_names,</span>
<span id="cb4-178"><a href="#cb4-178" aria-hidden="true" tabindex="-1"></a>                                      <span class="co">#target_metacell_size=...,</span></span>
<span id="cb4-179"><a href="#cb4-179" aria-hidden="true" tabindex="-1"></a>                                      random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb4-180"><a href="#cb4-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-181"><a href="#cb4-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-182"><a href="#cb4-182" aria-hidden="true" tabindex="-1"></a>metacells <span class="op">=</span> mc.pl.collect_metacells(clean, name<span class="op">=</span><span class="st">'cd34.metacells'</span>)</span></code></pre></div>
<p>Now we can save the results so we can load into R later.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># output dir</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>data_dir <span class="op">=</span> <span class="st">'./'</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># write the h5ad file</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>metacells.write_h5ad(<span class="st">'</span><span class="sc">{}</span><span class="st">/tutorial_MC2.h5ad'</span>.<span class="bu">format</span>(data_dir))</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># write obs/var tables</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>metacells.obs.to_csv(<span class="st">'</span><span class="sc">{}</span><span class="st">/tutorial_MC2_obs.csv'</span>.<span class="bu">format</span>(data_dir))</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>metacells.var.to_csv(<span class="st">'</span><span class="sc">{}</span><span class="st">/tutorial_MC2_var.csv'</span>.<span class="bu">format</span>(data_dir))</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co"># save the sparse matrix for Seurat:</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>X <span class="op">=</span> metacells.X</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>X <span class="op">=</span> scipy.sparse.csr_matrix(np.transpose(X).astype(<span class="bu">int</span>))</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>io.mmwrite(<span class="st">'</span><span class="sc">{}</span><span class="st">/tutorial_MC2.mtx'</span>.<span class="bu">format</span>(data_dir), X)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-the-seacells-and-mc2-metacell-datasets-into-r">Load the SEACells and MC2 metacell datasets into R<a class="anchor" aria-label="anchor" href="#load-the-seacells-and-mc2-metacell-datasets-into-r"></a>
</h2>
<p>Here we load the results from running SEACells and MC2 into R, and format the data as Seurat objects.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># single-cell analysis package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>

<span class="co"># plotting and data science packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span>

<span class="co"># co-expression network analysis packages:</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/" class="external-link">WGCNA</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://smorabit.github.io/hdWGCNA/">hdWGCNA</a></span><span class="op">)</span>

<span class="co"># network analysis &amp; visualization package:</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org" class="external-link">igraph</a></span><span class="op">)</span>

<span class="co"># using the cowplot theme for ggplot</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>

<span class="co"># set random seed for reproducibility</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span>

<span class="co"># location of the directory where the data was saved</span>
<span class="va">indir</span> <span class="op">&lt;-</span> <span class="st">'./'</span>; <span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="st">'./'</span>

<span class="co"># load the UMI counts gene expression matrix</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/externalFormats.html" class="external-link">readMM</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">indir</span>,<span class="st">'tutorial_singlecell.mtx'</span><span class="op">)</span><span class="op">)</span>

<span class="co"># load harmony matrix</span>
<span class="va">X_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">indir</span>, <span class="st">'tutorial_singlecell_pca.csv'</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">','</span>, header<span class="op">=</span><span class="cn">TRUE</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span>

<span class="co"># load the cell &amp; gene metadata table:</span>
<span class="va">cell_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">indir</span>, <span class="st">'tutorial_singlecell_obs.csv'</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">','</span>, header<span class="op">=</span><span class="cn">TRUE</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span>

<span class="va">gene_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">indir</span>, <span class="st">'tutorial_singlecell_var.csv'</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">','</span>, header<span class="op">=</span><span class="cn">TRUE</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span>

<span class="co"># get the umap from cell_meta:</span>
<span class="va">umap</span> <span class="op">&lt;-</span> <span class="va">cell_meta</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'UMAP_1'</span>, <span class="st">'UMAP_2'</span><span class="op">)</span><span class="op">]</span>

<span class="co"># set the rownames and colnames for the expression matrix:</span>
<span class="co"># for Seurat, rows of X are genes, cols of X are cells</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cell_meta</span><span class="op">)</span>
<span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_meta</span><span class="op">)</span>
<span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">X_pca</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cell_meta</span><span class="op">)</span>
<span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">umap</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cell_meta</span><span class="op">)</span>

<span class="co"># create a Seruat object:</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>
  counts <span class="op">=</span> <span class="va">X</span>,
  meta.data <span class="op">=</span> <span class="va">cell_meta</span>,
  assay <span class="op">=</span> <span class="st">"RNA"</span>,
  project <span class="op">=</span> <span class="st">"SEACells"</span>,
  min.features <span class="op">=</span> <span class="fl">0</span>,
  min.cells <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>

<span class="co"># set PCA reduction</span>
<span class="va">seurat_obj</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateDimReducObject.html" class="external-link">CreateDimReducObject</a></span><span class="op">(</span>
  embeddings <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">X_pca</span><span class="op">)</span>,
  key<span class="op">=</span><span class="st">"PC"</span>,
  assay<span class="op">=</span><span class="st">"RNA"</span>
<span class="op">)</span>

<span class="co"># set UMAP reduction</span>
<span class="va">seurat_obj</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">umap</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateDimReducObject.html" class="external-link">CreateDimReducObject</a></span><span class="op">(</span>
  embeddings <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">umap</span><span class="op">)</span>,
  key<span class="op">=</span><span class="st">"UMAP"</span>,
  assay<span class="op">=</span><span class="st">"RNA"</span>
<span class="op">)</span>

<span class="co"># normalize expression matrix</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># save data:</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">seurat_obj</span>, file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">'tutorial_seacells.rds'</span><span class="op">)</span><span class="op">)</span>

<span class="co">##############################################################</span>
<span class="co"># SEACells metacells</span>
<span class="co">##############################################################</span>

<span class="va">X</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/externalFormats.html" class="external-link">readMM</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">indir</span>,<span class="st">'tutorial_seacells.mtx'</span><span class="op">)</span><span class="op">)</span>

<span class="va">cell_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">indir</span>, <span class="st">'tutorial_seacells_obs.csv'</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">','</span>, header<span class="op">=</span><span class="cn">TRUE</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cell_meta</span><span class="op">)</span>
<span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_meta</span><span class="op">)</span>

<span class="co"># create a Seruat object:</span>
<span class="va">m_obj</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>
  counts <span class="op">=</span> <span class="va">X</span>,
  assay <span class="op">=</span> <span class="st">"RNA"</span>,
  project <span class="op">=</span> <span class="st">"SEACells"</span>,
  min.features <span class="op">=</span> <span class="fl">0</span>,
  min.cells <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">m_obj</span>, file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">'tutorial_seacells_metacell.rds'</span><span class="op">)</span><span class="op">)</span>

<span class="co">##############################################################</span>
<span class="co"># MC2 metacells</span>
<span class="co">##############################################################</span>

<span class="co"># load and type cast to sparse matrix</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/externalFormats.html" class="external-link">readMM</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">indir</span>,<span class="st">'tutorial_MC2.mtx'</span><span class="op">)</span><span class="op">)</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">X</span>, <span class="st">'dgCMatrix'</span><span class="op">)</span>

<span class="va">cell_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">indir</span>, <span class="st">'tutorial_MC2_obs.csv'</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">','</span>, header<span class="op">=</span><span class="cn">TRUE</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span>
<span class="va">gene_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">indir</span>, <span class="st">'tutorial_MC2_var.csv'</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">','</span>, header<span class="op">=</span><span class="cn">TRUE</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cell_meta</span><span class="op">)</span>
<span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_meta</span><span class="op">)</span>

<span class="co"># create a Seruat object:</span>
<span class="va">m_obj</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>
  counts <span class="op">=</span> <span class="va">X</span>,
  assay <span class="op">=</span> <span class="st">"RNA"</span>,
  project <span class="op">=</span> <span class="st">"MC2"</span>,
  min.features <span class="op">=</span> <span class="fl">0</span>,
  min.cells <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">m_obj</span>, file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">'tutorial_MC2_metacell.rds'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Plot the UMAP and cluster assignments for the CD34+ HSC dataset:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_obj</span>, group.by<span class="op">=</span><span class="st">'celltype'</span>, label<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/umap_theme.html">umap_theme</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="figures/seacells/cd34_umap.png" width="600" height="900"></p>
</div>
<div class="section level2">
<h2 id="co-expression-network-analysis">Co-expression network analysis<a class="anchor" aria-label="anchor" href="#co-expression-network-analysis"></a>
</h2>
<p>Now we are ready to perform co-expression network analysis using these metacell datasets. First we need to load the CD34+ HSC dataset.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># directory where the data was saved</span>
 <span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="st">'./'</span>

<span class="co"># load CD34+ HSC dataset</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"tutorial_seacells.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="section level3">
<h3 id="seacells">SEACells<a class="anchor" aria-label="anchor" href="#seacells"></a>
</h3>
<p>Here we perform co-expression network analysis using the SEACells metacells. We can use the function <code>SetMetacellObject</code> to add the SEACells metacell data to the hdWGCNA experiment.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load datasets</span>
<span class="va">m_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">'tutorial_seacells_metacell.rds'</span><span class="op">)</span><span class="op">)</span>

<span class="co"># set up hdWGCNA experiment</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetupForWGCNA.html">SetupForWGCNA</a></span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  gene_select <span class="op">=</span> <span class="st">"fraction"</span>,
  fraction <span class="op">=</span> <span class="fl">0.05</span>,
  wgcna_name <span class="op">=</span> <span class="st">'SEACells'</span>
<span class="op">)</span>

<span class="co"># add the seacells dataset</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetMetacellObject.html">SetMetacellObject</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="va">m_obj</span><span class="op">)</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NormalizeMetacells.html">NormalizeMetacells</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># setup expression matrix</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetDatExpr.html">SetDatExpr</a></span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  group_name <span class="op">=</span> <span class="st">'all'</span>,
  use_metacells<span class="op">=</span><span class="cn">TRUE</span>,
<span class="op">)</span>

<span class="co"># test soft power threshold</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TestSoftPowers.html">TestSoftPowers</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># compute the co-expression network</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ConstructNetwork.html">ConstructNetwork</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># compute module eigengenes and eigengene-based connectivity</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleEigengenes.html">ModuleEigengenes</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleConnectivity.html">ModuleConnectivity</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># rename modules</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ResetModuleNames.html">ResetModuleNames</a></span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  new_name <span class="op">=</span> <span class="st">'sc-M'</span>,
  wgcna_name<span class="op">=</span><span class="st">'SEACells'</span>
<span class="op">)</span>

<span class="co"># plot module eigengenes</span>
<span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleFeaturePlot.html">ModuleFeaturePlot</a></span><span class="op">(</span><span class="va">seurat_obj</span>, order<span class="op">=</span><span class="cn">TRUE</span>, raster<span class="op">=</span><span class="cn">TRUE</span>, alpha<span class="op">=</span><span class="fl">1</span>, restrict_range<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># assemble plots</span>
<span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">6</span><span class="op">)</span></code></pre></div>
<p><img src="figures/seacells/featureplot_MEs_seacells.png" width="800" height="900"></p>
</div>
<div class="section level3">
<h3 id="mc2">MC2<a class="anchor" aria-label="anchor" href="#mc2"></a>
</h3>
<p>Here we perform co-expression network analysis using the MC2 metacells. This is nearly the same as above, but here we add an extra step to ensure that all of the genes selected for WGCNA are in the MC2 metacell object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">m_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">'tutorial_MC2_metacell.rds'</span><span class="op">)</span><span class="op">)</span>

<span class="co"># set up hdWGCNA experiment</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetupForWGCNA.html">SetupForWGCNA</a></span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  gene_select <span class="op">=</span> <span class="st">"fraction"</span>,
  fraction <span class="op">=</span> <span class="fl">0.05</span>,
  wgcna_name <span class="op">=</span> <span class="st">'MC2'</span>
<span class="op">)</span>

<span class="co"># IMPORTANT:</span>
<span class="co"># in the MC2 code above, we had to exclude some of the genes, so here we have to</span>
<span class="co"># make sure the genes that we selected for WGCNA are actually in the metacell</span>
<span class="co"># dataset</span>
<span class="va">wgcna_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetWGCNAGenes.html">GetWGCNAGenes</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>
<span class="va">wgcna_genes</span> <span class="op">&lt;-</span> <span class="va">wgcna_genes</span><span class="op">[</span><span class="va">wgcna_genes</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">m_obj</span><span class="op">)</span><span class="op">]</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetWGCNAGenes.html">SetWGCNAGenes</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="va">wgcna_genes</span><span class="op">)</span>

<span class="co"># add the MC2 dataset</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetMetacellObject.html">SetMetacellObject</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="va">m_obj</span><span class="op">)</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NormalizeMetacells.html">NormalizeMetacells</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># setup expression matrix</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetDatExpr.html">SetDatExpr</a></span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  group_name <span class="op">=</span> <span class="st">'all'</span>,
  use_metacells<span class="op">=</span><span class="cn">TRUE</span>,
<span class="op">)</span>

<span class="co"># test soft power threshold</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TestSoftPowers.html">TestSoftPowers</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># compute the co-expression network</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ConstructNetwork.html">ConstructNetwork</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># compute module eigengenes and eigengene-based connectivity</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleEigengenes.html">ModuleEigengenes</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleConnectivity.html">ModuleConnectivity</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># rename modules</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ResetModuleNames.html">ResetModuleNames</a></span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  new_name <span class="op">=</span> <span class="st">'mc2-M'</span>,
  wgcna_name<span class="op">=</span><span class="st">'MC2'</span>
<span class="op">)</span>

<span class="co"># plot module eigengenes</span>
<span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleFeaturePlot.html">ModuleFeaturePlot</a></span><span class="op">(</span><span class="va">seurat_obj</span>, order<span class="op">=</span><span class="cn">TRUE</span>, raster<span class="op">=</span><span class="cn">TRUE</span>, alpha<span class="op">=</span><span class="fl">1</span>, restrict_range<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># assemble plots</span>
<span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">6</span><span class="op">)</span></code></pre></div>
<p><img src="figures/seacells/featureplot_MEs_MC2.png" width="800" height="900"></p>
</div>
<div class="section level3">
<h3 id="hdwgcna">hdWGCNA<a class="anchor" aria-label="anchor" href="#hdwgcna"></a>
</h3>
<p>Lastly, we run the hdWGCNA metacell algorithm so we can compare the results of the three different methods. Since MC2 and SEACells don’t aggregate metacells separately by cluster or biological replicate, here we run hdWGCNA in the same manner where metacells are constructed for the whole dataset. This means that some metacells will be a mix of different cell types, which is also the case in SEACells and MC2.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># set up hdWGCNA experiment</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetupForWGCNA.html">SetupForWGCNA</a></span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  gene_select <span class="op">=</span> <span class="st">"fraction"</span>,
  fraction <span class="op">=</span> <span class="fl">0.05</span>,
  wgcna_name <span class="op">=</span> <span class="st">'hdWGCNA'</span>
<span class="op">)</span>

<span class="co"># set up dummy variable so we can run MetacellsByGroups for all clusters together</span>
<span class="va">seurat_obj</span><span class="op">$</span><span class="va">all_cells</span> <span class="op">&lt;-</span> <span class="st">'all'</span>

<span class="co"># run hdWGCNA metacell aggregation</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MetacellsByGroups.html">MetacellsByGroups</a></span><span class="op">(</span>
  seurat_obj <span class="op">=</span> <span class="va">seurat_obj</span>,
  group.by <span class="op">=</span> <span class="st">"all_cells"</span>,
  k <span class="op">=</span> <span class="fl">50</span>,
  target_metacells<span class="op">=</span><span class="fl">250</span>,
  ident.group <span class="op">=</span> <span class="st">'all_cells'</span>,
  min_cells<span class="op">=</span><span class="fl">0</span>,
  max_shared<span class="op">=</span><span class="fl">5</span>,
<span class="op">)</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NormalizeMetacells.html">NormalizeMetacells</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># setup expression matrix</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetDatExpr.html">SetDatExpr</a></span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  group_name <span class="op">=</span> <span class="st">'all'</span>,
  use_metacells<span class="op">=</span><span class="cn">TRUE</span>,
<span class="op">)</span>


<span class="co"># test soft power threshold</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TestSoftPowers.html">TestSoftPowers</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># compute the co-expression network</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ConstructNetwork.html">ConstructNetwork</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># compute module eigengenes and eigengene-based connectivity</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleEigengenes.html">ModuleEigengenes</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleConnectivity.html">ModuleConnectivity</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span>

<span class="co"># rename modules</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ResetModuleNames.html">ResetModuleNames</a></span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  new_name <span class="op">=</span> <span class="st">'hd-M'</span>,
  wgcna_name<span class="op">=</span><span class="st">'hdWGCNA'</span>
<span class="op">)</span>

<span class="co"># plot module eigengenes</span>
<span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleFeaturePlot.html">ModuleFeaturePlot</a></span><span class="op">(</span><span class="va">seurat_obj</span>, order<span class="op">=</span><span class="cn">TRUE</span>, raster<span class="op">=</span><span class="cn">TRUE</span>, alpha<span class="op">=</span><span class="fl">1</span>, restrict_range<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># assemble plots</span>
<span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">6</span><span class="op">)</span></code></pre></div>
<p><img src="figures/seacells/featureplot_MEs_hdWGCNA.png" width="800" height="900"></p>
</div>
</div>
<div class="section level2">
<h2 id="comparing-co-expression-modules-across-methods">Comparing co-expression modules across methods<a class="anchor" aria-label="anchor" href="#comparing-co-expression-modules-across-methods"></a>
</h2>
<p>For this CD34+ HSC dataset, each of the three metacell methods resulted in a set of gene modules, and the module eigengene FeaturePlots look like there is cell-type/lineage specificity of these modules. In this section, we will compare the results of these different analyses to get an idea of what is shared and distinct.</p>
<div class="section level3">
<h3 id="compare-dendrograms-and-module-assignments">Compare dendrograms and module assignments<a class="anchor" aria-label="anchor" href="#compare-dendrograms-and-module-assignments"></a>
</h3>
<p>We can plot the hdWGCNA dendrogram with the gene module color assignments below for all three methods as a high-level comparison of the different approaches.</p>
<details><summary>
SEACells dendrogram code
</summary><div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">m1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetModules.html">GetModules</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">'SEACells'</span><span class="op">)</span>
<span class="va">m2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetModules.html">GetModules</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">'hdWGCNA'</span><span class="op">)</span>
<span class="va">m3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetModules.html">GetModules</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">'MC2'</span><span class="op">)</span>

<span class="co"># get WGCNA network and module data</span>
<span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetNetworkData.html">GetNetworkData</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">"SEACells"</span><span class="op">)</span>

<span class="va">m1_genes</span> <span class="op">&lt;-</span> <span class="va">m1</span><span class="op">$</span><span class="va">gene_name</span>
<span class="va">m1_colors</span> <span class="op">&lt;-</span> <span class="va">m1</span><span class="op">$</span><span class="va">color</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m1_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m1</span><span class="op">$</span><span class="va">gene_name</span>

<span class="va">m2_colors</span> <span class="op">&lt;-</span> <span class="va">m2</span><span class="op">[</span><span class="va">m1</span><span class="op">$</span><span class="va">gene_name</span>, <span class="st">'color'</span><span class="op">]</span>
<span class="va">m2_colors</span><span class="op">[</span><span class="va">m2_colors</span> <span class="op">==</span> <span class="cn">NA</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'grey'</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m2_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m1</span><span class="op">$</span><span class="va">gene_name</span>

<span class="va">m3_colors</span> <span class="op">&lt;-</span> <span class="va">m3</span><span class="op">[</span><span class="va">m1</span><span class="op">$</span><span class="va">gene_name</span>, <span class="st">'color'</span><span class="op">]</span>
<span class="va">m3_colors</span><span class="op">[</span><span class="va">m3_colors</span> <span class="op">==</span> <span class="cn">NA</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'grey'</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m3_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m1</span><span class="op">$</span><span class="va">gene_name</span>

<span class="va">color_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>
  SEACells <span class="op">=</span> <span class="va">m1_colors</span>,
  hdWGCNA <span class="op">=</span> <span class="va">m2_colors</span>,
  MC2 <span class="op">=</span> <span class="va">m3_colors</span>
<span class="op">)</span>

<span class="co"># plot dendrogram</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">"compare_dendro_sc.png"</span><span class="op">)</span>,height<span class="op">=</span><span class="fl">3</span>, width<span class="op">=</span><span class="fl">5</span>, res<span class="op">=</span><span class="fl">600</span>, units<span class="op">=</span><span class="st">'in'</span><span class="op">)</span>
<span class="fu">WGCNA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/plotDendroAndColors.html" class="external-link">plotDendroAndColors</a></span><span class="op">(</span>
  <span class="va">net</span><span class="op">$</span><span class="va">dendrograms</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,
  <span class="va">color_df</span>,
  groupLabels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">color_df</span><span class="op">)</span>,
  dendroLabels <span class="op">=</span> <span class="cn">FALSE</span>, hang <span class="op">=</span> <span class="fl">0.03</span>, addGuide <span class="op">=</span> <span class="cn">TRUE</span>, guideHang <span class="op">=</span> <span class="fl">0.05</span>,
  main <span class="op">=</span> <span class="st">"SEACells dendro"</span>,
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</details><p><img src="figures/seacells/compare_dendro_sc.png" width="800" height="900"></p>
<details><summary>
MC2 dendrogram code
</summary><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">m1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetModules.html">GetModules</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">'SEACells'</span><span class="op">)</span>
<span class="va">m2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetModules.html">GetModules</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">'hdWGCNA'</span><span class="op">)</span>
<span class="va">m3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetModules.html">GetModules</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">'MC2'</span><span class="op">)</span>

<span class="co"># get WGCNA network and module data</span>
<span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetNetworkData.html">GetNetworkData</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">"MC2"</span><span class="op">)</span>


<span class="va">m3_genes</span> <span class="op">&lt;-</span> <span class="va">m3</span><span class="op">$</span><span class="va">gene_name</span>
<span class="va">m3_colors</span> <span class="op">&lt;-</span> <span class="va">m3</span><span class="op">$</span><span class="va">color</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m3_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m3</span><span class="op">$</span><span class="va">gene_name</span>

<span class="va">m2_colors</span> <span class="op">&lt;-</span> <span class="va">m2</span><span class="op">[</span><span class="va">m3</span><span class="op">$</span><span class="va">gene_name</span>, <span class="st">'color'</span><span class="op">]</span>
<span class="va">m2_colors</span><span class="op">[</span><span class="va">m2_colors</span> <span class="op">==</span> <span class="cn">NA</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'grey'</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m2_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m3</span><span class="op">$</span><span class="va">gene_name</span>

<span class="va">m1_colors</span> <span class="op">&lt;-</span> <span class="va">m1</span><span class="op">[</span><span class="va">m3</span><span class="op">$</span><span class="va">gene_name</span>, <span class="st">'color'</span><span class="op">]</span>
<span class="va">m1_colors</span><span class="op">[</span><span class="va">m1_colors</span> <span class="op">==</span> <span class="cn">NA</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'grey'</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m1_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m3</span><span class="op">$</span><span class="va">gene_name</span>


<span class="va">color_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>
  MC2 <span class="op">=</span> <span class="va">m3_colors</span>,
  hdWGCNA <span class="op">=</span> <span class="va">m2_colors</span>,
  SEACells <span class="op">=</span> <span class="va">m1_colors</span>

<span class="op">)</span>

<span class="co"># plot dendrogram</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">"compare_dendro_mc2.png"</span><span class="op">)</span>,height<span class="op">=</span><span class="fl">3</span>, width<span class="op">=</span><span class="fl">5</span>, res<span class="op">=</span><span class="fl">600</span>, units<span class="op">=</span><span class="st">'in'</span><span class="op">)</span>
<span class="fu">WGCNA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/plotDendroAndColors.html" class="external-link">plotDendroAndColors</a></span><span class="op">(</span>
  <span class="va">net</span><span class="op">$</span><span class="va">dendrograms</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,
  <span class="va">color_df</span>,
  groupLabels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">color_df</span><span class="op">)</span>,
  dendroLabels <span class="op">=</span> <span class="cn">FALSE</span>, hang <span class="op">=</span> <span class="fl">0.03</span>, addGuide <span class="op">=</span> <span class="cn">TRUE</span>, guideHang <span class="op">=</span> <span class="fl">0.05</span>,
  main <span class="op">=</span> <span class="st">"MC2 dendro"</span>,
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</details><p><img src="figures/seacells/compare_dendro_mc2.png" width="800" height="900"></p>
<details><summary>
hdWGCNA dendrogram code
</summary><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">m1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetModules.html">GetModules</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">'SEACells'</span><span class="op">)</span>
<span class="va">m2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetModules.html">GetModules</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">'hdWGCNA'</span><span class="op">)</span>
<span class="va">m3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetModules.html">GetModules</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">'MC2'</span><span class="op">)</span>

<span class="co"># get WGCNA network and module data</span>
<span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetNetworkData.html">GetNetworkData</a></span><span class="op">(</span><span class="va">seurat_obj</span>, wgcna_name<span class="op">=</span><span class="st">"hdWGCNA"</span><span class="op">)</span>


<span class="va">m2_genes</span> <span class="op">&lt;-</span> <span class="va">m2</span><span class="op">$</span><span class="va">gene_name</span>

<span class="va">m2_colors</span> <span class="op">&lt;-</span> <span class="va">m2</span><span class="op">$</span><span class="va">color</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m2_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m2</span><span class="op">$</span><span class="va">gene_name</span>

<span class="va">m1_colors</span> <span class="op">&lt;-</span> <span class="va">m1</span><span class="op">[</span><span class="va">m2</span><span class="op">$</span><span class="va">gene_name</span>, <span class="st">'color'</span><span class="op">]</span>
<span class="va">m1_colors</span><span class="op">[</span><span class="va">m2_colors</span> <span class="op">==</span> <span class="cn">NA</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'grey'</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m1_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m1</span><span class="op">$</span><span class="va">gene_name</span>

<span class="va">m3_colors</span> <span class="op">&lt;-</span> <span class="va">m3</span><span class="op">[</span><span class="va">m2</span><span class="op">$</span><span class="va">gene_name</span>, <span class="st">'color'</span><span class="op">]</span>
<span class="va">m3_colors</span><span class="op">[</span><span class="va">m2_colors</span> <span class="op">==</span> <span class="cn">NA</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'grey'</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m3_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m1</span><span class="op">$</span><span class="va">gene_name</span>


<span class="va">color_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>
  hdWGCNA <span class="op">=</span> <span class="va">m2_colors</span>,
  MC2 <span class="op">=</span> <span class="va">m3_colors</span>,
  SEACells <span class="op">=</span> <span class="va">m1_colors</span>
<span class="op">)</span>

<span class="co"># plot dendrogram</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">"compare_dendro_hdWGCNA.png"</span><span class="op">)</span>,height<span class="op">=</span><span class="fl">3</span>, width<span class="op">=</span><span class="fl">5</span>, res<span class="op">=</span><span class="fl">600</span>, units<span class="op">=</span><span class="st">'in'</span><span class="op">)</span>
<span class="fu">WGCNA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/plotDendroAndColors.html" class="external-link">plotDendroAndColors</a></span><span class="op">(</span>
  <span class="va">net</span><span class="op">$</span><span class="va">dendrograms</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,
  <span class="va">color_df</span>,
  groupLabels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">color_df</span><span class="op">)</span>,
  dendroLabels <span class="op">=</span> <span class="cn">FALSE</span>, hang <span class="op">=</span> <span class="fl">0.03</span>, addGuide <span class="op">=</span> <span class="cn">TRUE</span>, guideHang <span class="op">=</span> <span class="fl">0.05</span>,
  main <span class="op">=</span> <span class="st">"hdWGCNA dendro"</span>,
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</details><p><img src="figures/seacells/compare_dendro_hdWGCNA.png" width="800" height="900"></p>
</div>
<div class="section level3">
<h3 id="gene-overlap-analysis">Gene overlap analysis<a class="anchor" aria-label="anchor" href="#gene-overlap-analysis"></a>
</h3>
<p>Here we perform statistical tests to compare the overlap between the sets of co-expression modules that we got from the three different metacell approaches.</p>
<p><strong><em>This section of the tutorial is under construction</em></strong></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
