<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Cross-species and cross-modality analysis • hdWGCNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Cross-species and cross-modality analysis">
<meta name="description" content="Projecting hdWGCNA co-expression modules across species and -omics modalities.
">
<meta property="og:description" content="Projecting hdWGCNA co-expression modules across species and -omics modalities.
">
<meta property="og:image" content="https://smorabit.github.io/hdWGCNA/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">hdWGCNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.00</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/basic_tutorial.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><h6 class="dropdown-header" data-toc-skip>Core functionality</h6></li>
    <li><a class="dropdown-item" href="../articles/basic_tutorial.html">hdWGCNA in single-cell data</a></li>
    <li><a class="dropdown-item" href="../articles/ST_basics.html">hdWGCNA in spatial transcriptomics data</a></li>
    <li><a class="dropdown-item" href="../articles/isoform_pbmcs.html">Isoform co-expression network analysis with PacBio MAS-Seq</a></li>
    <li><a class="dropdown-item" href="../articles/network_visualizations.html">Network visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Biological context for co-expression modules</h6></li>
    <li><a class="dropdown-item" href="../articles/differential_MEs.html">Differential module eigengene (DME) analysis</a></li>
    <li><a class="dropdown-item" href="../articles/module_trait_correlation.html">Module trait correlation</a></li>
    <li><a class="dropdown-item" href="../articles/enrichment_analysis.html">Enrichment analysis</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Exploring modules in external datasets</h6></li>
    <li><a class="dropdown-item" href="../articles/projecting_modules.html">Projecting modules to new datasets</a></li>
    <li><a class="dropdown-item" href="../articles/module_preservation.html">Module preservation and reproducibility</a></li>
    <li><a class="dropdown-item" href="../articles/projecting_modules_cross.html">Cross-species and cross-modality analysis</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Advanced topics</h6></li>
    <li><a class="dropdown-item" href="../articles/tf_network.html">Transcription factor regulatory network analysis</a></li>
    <li><a class="dropdown-item" href="../articles/consensus_wgcna.html">Consensus network analysis</a></li>
    <li><a class="dropdown-item" href="../articles/pseudobulk.html">hdWGCNA with pseudobulk data</a></li>
    <li><a class="dropdown-item" href="../articles/pseudotime.html">Co-expression module dynamics with pseudotime</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Other</h6></li>
    <li><a class="dropdown-item" href="../articles/customization.html">Module customization</a></li>
    <li><a class="dropdown-item" href="../articles/other_metacells.html">Alternative metacell algorithms</a></li>
    <li><a class="dropdown-item" href="../articles/sctransform.html">Using SCTransform normalized data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">All vignettes</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/smorabit/hdWGCNA/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Cross-species and cross-modality analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/smorabit/hdWGCNA/blob/HEAD/vignettes/projecting_modules_cross.Rmd" class="external-link"><code>vignettes/projecting_modules_cross.Rmd</code></a></small>
      <div class="d-none name"><code>projecting_modules_cross.Rmd</code></div>
    </div>

    
    
<p>In this tutorial, we demonstrate how to use hdWGCNA to project co-expression modules from a reference to a query dataset in certain special cases. Specifically, we project modules from a human snRNA-seq dataset to a human snATAC-seq dataset, a mouse snRNA-seq dataset, and a mouse <a href="https://www.10xgenomics.com/products/spatial-gene-expression" class="external-link">Visium</a> spatial transcriptomics dataset.</p>
<p>Load the snRNA-seq data and the required libraries:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># single-cell analysis package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plotting and data science packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># co-expressi fron network analysis packages:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/" class="external-link">WGCNA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://smorabit.github.io/hdWGCNA/">hdWGCNA</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># network analysis &amp; visualization package:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># using the cowplot theme for ggplot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set random seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load the Zhou et al snRNA-seq dataset</span></span>
<span><span class="va">seurat_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'data/Zhou_control.rds'</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="project-modules-from-rna-seq-to-atac-seq">Project modules from RNA-seq to ATAC-seq<a class="anchor" aria-label="anchor" href="#project-modules-from-rna-seq-to-atac-seq"></a>
</h2>
<p>In this section, we project snRNA-seq co-expression modules onto the snATAC-seq dataset from our <a href="https://doi.org/10.1038/s41588-021-00894-z" class="external-link">Nature Genetics paper</a>. This analysis requires a fully processed snRNA-seq dataset that we have run hdWGCNA on, and a fully processed snATAC-seq dataset. For snATAC-seq processing and data formatting, we suggest using <a href="https://satijalab.org/signac/" class="external-link">Signac</a>, an extension of Seurat for performing single-cell chromatin accessibility analysis. Signac offers several <a href="https://satijalab.org/signac/articles/mouse_brain_vignette.html" class="external-link">tutorials</a> for processing snATAC-seq data from scratch, and before projecting co-expression modules into snATAC-seq data we need to do the following analysis steps:</p>
<ul>
<li>
<a href="https://satijalab.org/signac/articles/pbmc_vignette.html#computing-qc-metrics" class="external-link">Quality control and filtering</a>.</li>
<li>Term frequency inverse document frequency (TF-IDF) normalization (<a href="https://satijalab.org/signac/reference/runtfidf" class="external-link">RunTFIDF</a>).</li>
<li>Feature Selection (<a href="https://satijalab.org/signac/reference/findtopfeatures" class="external-link">FindTopFeatures</a>).</li>
<li>Linear dimensionality reduction (<a href="https://satijalab.org/signac/reference/runsvd" class="external-link">RunSVD</a>).</li>
<li>Non-linear dimensionality reduction (<a href="https://satijalab.org/seurat/reference/runumap" class="external-link">RunUMAP</a>) for visualizations.</li>
<li>Group cells into clusters (<a href="https://satijalab.org/seurat/reference/findneighbors" class="external-link">FindNeighbors</a> and <a href="https://satijalab.org/seurat/reference/findclusters" class="external-link">FindClusters</a>).</li>
<li>Create a gene activity matrix (<a href="https://satijalab.org/signac/reference/geneactivity" class="external-link">GeneActivity</a>).</li>
</ul>
<div class="section level3">
<h3 id="load-snatac-seq-dataset">Load snATAC-seq dataset<a class="anchor" aria-label="anchor" href="#load-snatac-seq-dataset"></a>
</h3>
<p>This following code loads <code>Signac</code> and our snATAC-seq dataset into R.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># single-cell chromatin analysis package:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load dataset</span></span>
<span><span class="va">seurat_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'/dfs3b/swaruplab/smorabit/analysis/AD_NucSeq_2019/atac_analysis/all_data/celltype-analysis/data/NucSeq_processed_activity_qc_batch_correct.rds'</span><span class="op">)</span></span></code></pre></div>
<p>To ensure that we loaded the dataset, and that this is indeed a chromatin dataset, we plot the snATAC-seq UMAP and a chromatin accessibility coverage plot for <em>GAD2</em>, an inhibitory neuron marker gene.</p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot the snATAC-seq clusters:</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_atac</span>, group.by<span class="op">=</span><span class="st">'cell_type'</span>, label<span class="op">=</span><span class="cn">TRUE</span>, raster<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="../reference/umap_theme.html">umap_theme</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">CoveragePlot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">signac_atac</span>,</span>
<span>  group.by<span class="op">=</span><span class="st">'cell_type'</span>,</span>
<span>  region <span class="op">=</span> <span class="st">"GAD2"</span>,</span>
<span>  annotation <span class="op">=</span> <span class="cn">TRUE</span>, peaks <span class="op">=</span> <span class="cn">FALSE</span>, tile <span class="op">=</span> <span class="cn">FALSE</span>, links <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">'atac_umap_covplot.png'</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">12</span>, height<span class="op">=</span><span class="fl">4</span>, res<span class="op">=</span><span class="fl">400</span>, units<span class="op">=</span><span class="st">'in'</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>widths<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><p><img src="figures/projection/atac_umap_covplot.png" width="700" height="700"></p>
</div>
<div class="section level3">
<h3 id="project-modules">Project modules<a class="anchor" aria-label="anchor" href="#project-modules"></a>
</h3>
<p>In previous examples, the reference and query datasets were both transcriptomics datasets, but here we are trying to investigate co-expression modules in the epigenome using chromatin accessibility data. Here we use the <code>ProjectModules</code> function as we have done before, but we have to ensure that we are using the Gene Activity matrix from the snATAC-seq dataset rather than the chromatin accessibility matrix.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set assay to RNA (Gene Activity matrix)</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seurat_atac</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'RNA'</span></span>
<span></span>
<span><span class="co"># project modules</span></span>
<span><span class="va">seurat_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ProjectModules.html">ProjectModules</a></span><span class="op">(</span></span>
<span>  seurat_obj <span class="op">=</span> <span class="va">seurat_atac</span>,</span>
<span>  seurat_ref <span class="op">=</span> <span class="va">seurat_ref</span>,</span>
<span>  group.by.vars <span class="op">=</span> <span class="st">'Batch'</span>,</span>
<span>  wgcna_name <span class="op">=</span> <span class="st">"INH"</span>,</span>
<span>  wgcna_name_proj<span class="op">=</span><span class="st">"Zhou_projected"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute module hub scores for projected modules:</span></span>
<span><span class="va">seurat_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleExprScore.html">ModuleExprScore</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_atac</span>,</span>
<span>  n_genes <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  method<span class="op">=</span><span class="st">'Seurat'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># save the results:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">seurat_atac</span>, file<span class="op">=</span><span class="st">'mm2021_atac_hdWGCNA.rds'</span><span class="op">)</span></span></code></pre></div>
<p>Using the gene activity matrix we computed module eigengenes in the snATAC-seq dataset, harmonizing based on the sequencing batch. Additionally, we computed module hub gene expression scores for the projected modules, and now we are ready for visualization and downstream analysis.</p>
</div>
<div class="section level3">
<h3 id="visualize-modules-in-snatac-seq">Visualize modules in snATAC-seq<a class="anchor" aria-label="anchor" href="#visualize-modules-in-snatac-seq"></a>
</h3>
<p>Here we will visualize the hub gene expression scores of the projected modules in the snATAC-seq dataset using <code>ModuleFeaturePlot</code> and Seurat’s <code>DotPlot</code> function.</p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># only show a handful of selected modules</span></span>
<span><span class="va">selected_mods</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'INH-M'</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">7</span>,<span class="fl">8</span>,<span class="fl">9</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make a featureplot of hub scores for each module</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleFeaturePlot.html">ModuleFeaturePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_atac</span>,</span>
<span>  features<span class="op">=</span><span class="st">'scores'</span>,</span>
<span>  order<span class="op">=</span><span class="st">"shuffle"</span>,</span>
<span>  module_names <span class="op">=</span> <span class="va">selected_mods</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># stitch together with patchwork</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
</details><p><img src="figures/projection/atac_projected_featureplots_scores.png" width="700" height="700"></p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the projected hMEs</span></span>
<span><span class="va">mod_scores</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/GetModuleScores.html">GetModuleScores</a></span><span class="op">(</span><span class="va">seurat_atac</span><span class="op">)</span></span>
<span><span class="va">mod_scores</span> <span class="op">&lt;-</span> <span class="va">mod_scores</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mod_scores</span><span class="op">)</span> <span class="op">!=</span> <span class="st">'grey'</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># add hMEs to Seurat meta-data:</span></span>
<span><span class="va">seurat_atac</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_atac</span><span class="op">@</span><span class="va">meta.data</span>,</span>
<span>  <span class="va">mod_scores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot with Seurat's DotPlot function</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_atac</span>,</span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mod_scores</span><span class="op">)</span>,</span>
<span>    group.by <span class="op">=</span> <span class="st">'cell_type'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># flip the x/y axes, rotate the axis labels, and change color scheme:</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">RotatedAxis</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span>high<span class="op">=</span><span class="st">'darkorchid4'</span>, mid<span class="op">=</span><span class="st">'grey95'</span>, low<span class="op">=</span><span class="st">'seagreen'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
</details><p><img src="figures/projection/atac_projected_scores_dotplot.png" width="700" height="700"></p>
</div>
<div class="section level3">
<h3 id="module-preservation-in-snatac-seq">Module preservation in snATAC-seq<a class="anchor" aria-label="anchor" href="#module-preservation-in-snatac-seq"></a>
</h3>
<p>In this section, we compute the module preservation statistics for the snATAC-seq dataset. This analysis will tell us which of the co-expression modules are conserved at the epigenome level, and specifically which network properties, such as density or connectiviy, are preserved across modalities. This code is very similar to our previous module preservation analysis, and we do not have to take any specific steps for snATAC-seq here. Note that this code can take a few hours to run on a large query dataset.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set dat expr for single-cell dataset:</span></span>
<span><span class="va">seurat_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetDatExpr.html">SetDatExpr</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_atac</span>,</span>
<span>  group_name <span class="op">=</span> <span class="st">"INH"</span>,</span>
<span>  group.by <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>  use_metacells <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run module preservation function</span></span>
<span><span class="va">seurat_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModulePreservation.html">ModulePreservation</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_atac</span>,</span>
<span>  seurat_ref <span class="op">=</span> <span class="va">seurat_ref</span>,</span>
<span>  name<span class="op">=</span><span class="st">"Zhou-INH"</span>,</span>
<span>  verbose<span class="op">=</span><span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># save the results again since this step takes a while to run:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">seurat_atac</span>, file<span class="op">=</span><span class="st">'mm2021_atac_hdWGCNA.rds'</span><span class="op">)</span></span></code></pre></div>
<p>Now we visualize the results using the <code>PlotModulePreservation</code> function.</p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot the summary stats</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotModulePreservation.html">PlotModulePreservation</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_atac</span>,</span>
<span>  name<span class="op">=</span><span class="st">"Zhou-INH"</span>,</span>
<span>  statistics <span class="op">=</span> <span class="st">"summary"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</details><p><img src="figures/projection/atac_module_preservation_summary.png" width="700" height="700"></p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot all of the stats togehter</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotModulePreservation.html">PlotModulePreservation</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_atac</span>,</span>
<span>  name<span class="op">=</span><span class="st">"Zhou-INH"</span>,</span>
<span>  statistics <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  plot_labels<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">6</span><span class="op">)</span></span></code></pre></div>
</details><p><img src="figures/projection/atac_module_preservation_all.png" width="700" height="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="project-modules-from-hg38-to-mm10">Project modules from hg38 to mm10<a class="anchor" aria-label="anchor" href="#project-modules-from-hg38-to-mm10"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load Zhou et al 5xFAD dataset:</span></span>
<span><span class="va">seurat_mouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'/dfs3b/swaruplab/smorabit/analysis/AD_NucSeq_2019/batch_correction/liger/update/mouse_integration/data/zhou_5xFAD_ProcessedSeuratFinal.rds'</span><span class="op">)</span></span>
<span><span class="va">seurat_mouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">seurat_mouse</span>, <span class="va">Cell.Types</span> <span class="op">!=</span> <span class="st">'OTHER'</span><span class="op">)</span></span>
<span><span class="va">seurat_mouse</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="va">seurat_mouse</span><span class="op">$</span><span class="va">Cell.Types</span></span>
<span><span class="va">seurat_mouse</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">seurat_mouse</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span> <span class="op">==</span> <span class="fl">13</span>, <span class="st">'OPC'</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">seurat_mouse</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seurat_mouse</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">seurat_mouse</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">26</span>,<span class="fl">27</span><span class="op">)</span>, <span class="st">'PER'</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">seurat_mouse</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_mouse</span>, group.by <span class="op">=</span> <span class="st">'cell_type'</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/umap_theme.html">umap_theme</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'Zhou et al 5xFAD'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">'mouse_umap.png'</span><span class="op">)</span>, units<span class="op">=</span><span class="st">'in'</span>, res<span class="op">=</span><span class="fl">400</span>, width<span class="op">=</span><span class="fl">5</span>, height<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">p</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/projection/mouse_umap.png" width="500" height="500"></p>
<div class="section level3">
<h3 id="project-modules-to-5xfad-data">Project modules to 5xFAD data<a class="anchor" aria-label="anchor" href="#project-modules-to-5xfad-data"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load mouse &lt;-&gt; human gene name table:</span></span>
<span><span class="va">hg38_mm10_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span></span>
<span>  <span class="st">"/dfs3b/swaruplab/smorabit/resources/hg38_mm10_orthologs_2021.txt"</span>,</span>
<span>  sep<span class="op">=</span><span class="st">'\t'</span>,</span>
<span>  header<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">hg38_mm10_genes</span><span class="op">)</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'hg38_id'</span>, <span class="st">'mm10_id'</span>, <span class="st">'mm10_name'</span>, <span class="st">'hg38_name'</span><span class="op">)</span></span>
<span><span class="va">hg38_mm10_genes</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">hg38_mm10_genes</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">hg38_name</span>, <span class="va">mm10_name</span>, <span class="va">hg38_id</span>, <span class="va">mm10_id</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">seurat_mouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ProjectModules.html">ProjectModules</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_mouse</span>,</span>
<span>  seurat_ref <span class="op">=</span> <span class="va">seurat_ref</span>,</span>
<span>  gene_mapping<span class="op">=</span><span class="va">hg38_mm10_genes</span>,</span>
<span>  genome1_col<span class="op">=</span><span class="st">"hg38_name"</span>, <span class="co"># genome of reference data</span></span>
<span>  genome2_col<span class="op">=</span><span class="st">"mm10_name"</span>, <span class="co"># genome of query data</span></span>
<span>  wgcna_name_proj <span class="op">=</span> <span class="st">"Zhou-INH"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute module hub scores for projected modules:</span></span>
<span><span class="va">seurat_mouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleExprScore.html">ModuleExprScore</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_mouse</span>,</span>
<span>  n_genes <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  method<span class="op">=</span><span class="st">'Seurat'</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-modules-in-5xfad-snrna-seq">Visualize modules in 5xFAD snRNA-seq<a class="anchor" aria-label="anchor" href="#visualize-modules-in-5xfad-snrna-seq"></a>
</h3>
<p>Visualize FeaturePlots</p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># only show a handful of selected modules</span></span>
<span><span class="va">selected_mods</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'INH-M'</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">7</span>,<span class="fl">8</span>,<span class="fl">9</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make a featureplot of hub scores for each module</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleFeaturePlot.html">ModuleFeaturePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_mouse</span>,</span>
<span>  features<span class="op">=</span><span class="st">'scores'</span>,</span>
<span>  order<span class="op">=</span><span class="st">"order"</span>,</span>
<span>  module_names <span class="op">=</span> <span class="va">selected_mods</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># stitch together with patchwork</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">'mouse_featureplot.png'</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">12</span>, height<span class="op">=</span><span class="fl">6</span>, units<span class="op">=</span><span class="st">'in'</span>, res<span class="op">=</span><span class="fl">400</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><p><img src="figures/projection/mouse_featureplot.png" width="700" height="700"></p>
<p>Visualize DotPlots</p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the projected hMEs</span></span>
<span><span class="va">mod_scores</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/GetModuleScores.html">GetModuleScores</a></span><span class="op">(</span><span class="va">seurat_mouse</span><span class="op">)</span></span>
<span><span class="va">mod_scores</span> <span class="op">&lt;-</span> <span class="va">mod_scores</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mod_scores</span><span class="op">)</span> <span class="op">!=</span> <span class="st">'grey'</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># add hMEs to Seurat meta-data:</span></span>
<span><span class="va">seurat_mouse</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_mouse</span><span class="op">@</span><span class="va">meta.data</span>,</span>
<span>  <span class="va">mod_scores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot with Seurat's DotPlot function</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_mouse</span>,</span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mod_scores</span><span class="op">)</span>,</span>
<span>    group.by <span class="op">=</span> <span class="st">'cell_type'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># flip the x/y axes, rotate the axis labels, and change color scheme:</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">RotatedAxis</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span>high<span class="op">=</span><span class="st">'red'</span>, mid<span class="op">=</span><span class="st">'grey95'</span>, low<span class="op">=</span><span class="st">'blue'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">'mouse_dotplot.png'</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">10</span>, height<span class="op">=</span><span class="fl">5</span>, units<span class="op">=</span><span class="st">'in'</span>, res<span class="op">=</span><span class="fl">400</span><span class="op">)</span></span>
<span><span class="va">p</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><p><img src="figures/projection/mouse_dotplot.png" width="700" height="700"></p>
</div>
<div class="section level3">
<h3 id="module-preservation">Module preservation<a class="anchor" aria-label="anchor" href="#module-preservation"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set dat expr for single-cell dataset:</span></span>
<span><span class="va">seurat_mouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetDatExpr.html">SetDatExpr</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_mouse</span>,</span>
<span>  group_name <span class="op">=</span> <span class="st">"INH"</span>,</span>
<span>  group.by <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>  use_metacells <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run module preservation function</span></span>
<span><span class="va">seurat_mouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModulePreservation.html">ModulePreservation</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_mouse</span>,</span>
<span>  seurat_ref <span class="op">=</span> <span class="va">seurat_ref</span>,</span>
<span>  name<span class="op">=</span><span class="st">"Zhou-INH"</span>,</span>
<span>  gene_mapping<span class="op">=</span><span class="va">hg38_mm10_genes</span>,</span>
<span>  genome1_col<span class="op">=</span><span class="st">"hg38_name"</span>, <span class="co"># genome of reference data</span></span>
<span>  genome2_col<span class="op">=</span><span class="st">"mm10_name"</span>, <span class="co"># genome of query data</span></span>
<span>  verbose<span class="op">=</span><span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">seurat_mouse</span>, file<span class="op">=</span><span class="st">'../data/5xFAD_zhou_hdWGCNA.rds'</span><span class="op">)</span></span></code></pre></div>
<p>Now we visualize the results using the <code>PlotModulePreservation</code> function.</p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot the summary stats</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotModulePreservation.html">PlotModulePreservation</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_mouse</span>,</span>
<span>  name<span class="op">=</span><span class="st">"Zhou-INH"</span>,</span>
<span>  statistics <span class="op">=</span> <span class="st">"summary"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">'mouse_preservation_summary.png'</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">10</span>, height<span class="op">=</span><span class="fl">5</span>, res<span class="op">=</span><span class="fl">400</span>, units<span class="op">=</span><span class="st">'in'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><p><img src="figures/projection/mouse_preservation_summary.png" width="700" height="700"></p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot all of the stats togehter</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotModulePreservation.html">PlotModulePreservation</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_mouse</span>,</span>
<span>  name<span class="op">=</span><span class="st">"Zhou-INH"</span>,</span>
<span>  statistics <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  plot_labels<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">'mouse_preservation_all.png'</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">20</span>, height<span class="op">=</span><span class="fl">20</span><span class="op">*</span><span class="op">(</span><span class="fl">2</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span>, res<span class="op">=</span><span class="fl">400</span>, units<span class="op">=</span><span class="st">'in'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><p><img src="figures/projection/visium_preservation_all.png" width="700" height="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="project-modules-to-visium-data">Project modules to visium data<a class="anchor" aria-label="anchor" href="#project-modules-to-visium-data"></a>
</h2>
<div class="section level3">
<h3 id="download-and-process-visium-brain-dataset">Download and process Visium brain dataset<a class="anchor" aria-label="anchor" href="#download-and-process-visium-brain-dataset"></a>
</h3>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load 10X genomics visium dataset</span></span>
<span><span class="fu">SeuratData</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">"stxBrain"</span><span class="op">)</span></span>
<span><span class="va">seurat_vis</span> <span class="op">&lt;-</span> <span class="fu">SeuratData</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"stxBrain"</span>, type <span class="op">=</span> <span class="st">"anterior1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># process the Visium dataset with Seurat:</span></span>
<span><span class="va">seurat_vis</span> <span class="op">&lt;-</span> <span class="va">seurat_vis</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>res<span class="op">=</span><span class="fl">0.75</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_vis</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/umap_theme.html">umap_theme</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialDimPlot</a></span><span class="op">(</span><span class="va">seurat_vis</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, label.size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">'vis_umap.png'</span><span class="op">)</span>, units<span class="op">=</span><span class="st">'in'</span>, res<span class="op">=</span><span class="fl">400</span>, width<span class="op">=</span><span class="fl">10</span>, height<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">|</span> <span class="va">p2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/projection/vis_umap.png" width="700" height="700"></p>
</div>
<div class="section level3">
<h3 id="project-modules-to-visium-dataset">Project modules to visium dataset:<a class="anchor" aria-label="anchor" href="#project-modules-to-visium-dataset"></a>
</h3>
<p>Load the mouse to human gene table</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load mouse &lt;-&gt; human gene name table:</span></span>
<span><span class="va">hg38_mm10_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span></span>
<span>  <span class="st">"/dfs3b/swaruplab/smorabit/resources/hg38_mm10_orthologs_2021.txt"</span>,</span>
<span>  sep<span class="op">=</span><span class="st">'\t'</span>,</span>
<span>  header<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">hg38_mm10_genes</span><span class="op">)</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'hg38_id'</span>, <span class="st">'mm10_id'</span>, <span class="st">'mm10_name'</span>, <span class="st">'hg38_name'</span><span class="op">)</span></span>
<span><span class="va">hg38_mm10_genes</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">hg38_mm10_genes</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">hg38_name</span>, <span class="va">mm10_name</span>, <span class="va">hg38_id</span>, <span class="va">mm10_id</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Project Modules</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ProjectModules.html">ProjectModules</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_vis</span>,</span>
<span>  seurat_ref <span class="op">=</span> <span class="va">seurat_ref</span>,</span>
<span>  gene_mapping<span class="op">=</span><span class="va">hg38_mm10_genes</span>,</span>
<span>  genome1_col<span class="op">=</span><span class="st">"hg38_name"</span>, <span class="co"># genome of reference data</span></span>
<span>  genome2_col<span class="op">=</span><span class="st">"mm10_name"</span>, <span class="co"># genome of query data</span></span>
<span>  wgcna_name_proj <span class="op">=</span> <span class="st">"Zhou-INH"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute module hub scores for projected modules:</span></span>
<span><span class="va">seurat_vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleExprScore.html">ModuleExprScore</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_vis</span>,</span>
<span>  n_genes <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  method<span class="op">=</span><span class="st">'Seurat'</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Spatial Feature Plot:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get all of the image coordinates for BayesSpace</span></span>
<span><span class="va">image_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">seurat_vis</span><span class="op">@</span><span class="va">images</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">cur_image</span><span class="op">)</span><span class="op">{</span><span class="va">seurat_vis</span><span class="op">@</span><span class="va">images</span><span class="op">[[</span><span class="va">cur_image</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">coordinates</span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># re-order the rows of image_df to match the seurat_obj</span></span>
<span><span class="va">image_df</span> <span class="op">&lt;-</span> <span class="va">image_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">seurat_vis</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/all.equal-methods.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">image_df</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">seurat_vis</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make a spatial "reduction"</span></span>
<span><span class="va">image_emb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">image_df</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'col'</span>, <span class="st">'row'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">image_emb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Spatial_1"</span>, <span class="st">"Spatial_2"</span><span class="op">)</span></span>
<span><span class="va">seurat_vis</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateDimReducObject.html" class="external-link">CreateDimReducObject</a></span><span class="op">(</span></span>
<span>  <span class="va">image_emb</span>,</span>
<span>  key <span class="op">=</span> <span class="st">'Spatial'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># only show a handful of selected modules</span></span>
<span><span class="va">selected_mods</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'INH-M'</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">7</span>,<span class="fl">8</span>,<span class="fl">9</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make a featureplot of hub scores for each module</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleFeaturePlot.html">ModuleFeaturePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_vis</span>,</span>
<span>  features<span class="op">=</span><span class="st">'scores'</span>,</span>
<span>  order<span class="op">=</span><span class="st">"order"</span>,</span>
<span>  module_names <span class="op">=</span> <span class="va">selected_mods</span>,</span>
<span>  reduction<span class="op">=</span><span class="st">'spatial'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># stitch together with patchwork</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">'visium_featureplot.png'</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">12</span>, height<span class="op">=</span><span class="fl">6</span>, units<span class="op">=</span><span class="st">'in'</span>, res<span class="op">=</span><span class="fl">400</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/projection/visium_featureplot.png" width="700" height="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make a featureplot of hub scores for each module</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleFeaturePlot.html">ModuleFeaturePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_vis</span>,</span>
<span>  features<span class="op">=</span><span class="st">'scores'</span>,</span>
<span>  order<span class="op">=</span><span class="st">"order"</span>,</span>
<span>  module_names <span class="op">=</span> <span class="va">selected_mods</span>,</span>
<span>  reduction<span class="op">=</span><span class="st">'umap'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># stitch together with patchwork</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">'visium_featureplot_umap.png'</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">12</span>, height<span class="op">=</span><span class="fl">6</span>, units<span class="op">=</span><span class="st">'in'</span>, res<span class="op">=</span><span class="fl">400</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/projection/visium_featureplot_umap.png" width="700" height="700"></p>
</div>
<div class="section level3">
<h3 id="module-preservation-in-visium">Module preservation in Visium:<a class="anchor" aria-label="anchor" href="#module-preservation-in-visium"></a>
</h3>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set dat expr for single-cell dataset:</span></span>
<span><span class="va">seurat_vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetDatExpr.html">SetDatExpr</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_vis</span>,</span>
<span>  use_metacells <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run module preservation function</span></span>
<span><span class="va">seurat_vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModulePreservation.html">ModulePreservation</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_vis</span>,</span>
<span>  seurat_ref <span class="op">=</span> <span class="va">seurat_ref</span>,</span>
<span>  name<span class="op">=</span><span class="st">"Zhou-INH"</span>,</span>
<span>  verbose<span class="op">=</span><span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModulePreservation.html">ModulePreservation</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_vis</span>,</span>
<span>  seurat_ref <span class="op">=</span> <span class="va">seurat_ref</span>,</span>
<span>  name<span class="op">=</span><span class="st">"Zhou-INH"</span>,</span>
<span>  gene_mapping<span class="op">=</span><span class="va">hg38_mm10_genes</span>,</span>
<span>  genome1_col<span class="op">=</span><span class="st">"hg38_name"</span>, <span class="co"># genome of reference data</span></span>
<span>  genome2_col<span class="op">=</span><span class="st">"mm10_name"</span>, <span class="co"># genome of query data</span></span>
<span>  verbose<span class="op">=</span><span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">seurat_vis</span>, file<span class="op">=</span><span class="st">'../data/10x_visium_hdWGCNA.rds'</span><span class="op">)</span></span></code></pre></div>
<p>Now we visualize the results using the <code>PlotModulePreservation</code> function.</p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot the summary stats</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotModulePreservation.html">PlotModulePreservation</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_vis</span>,</span>
<span>  name<span class="op">=</span><span class="st">"Zhou-INH"</span>,</span>
<span>  statistics <span class="op">=</span> <span class="st">"summary"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">'visium_preservation_summary.png'</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">10</span>, height<span class="op">=</span><span class="fl">5</span>, res<span class="op">=</span><span class="fl">400</span>, units<span class="op">=</span><span class="st">'in'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><p><img src="figures/projection/visium_preservation_summary.png" width="700" height="700"></p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot all of the stats togehter</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotModulePreservation.html">PlotModulePreservation</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_vis</span>,</span>
<span>  name<span class="op">=</span><span class="st">"Zhou-INH"</span>,</span>
<span>  statistics <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  plot_labels<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">fig_dir</span>, <span class="st">'visium_preservation_all.png'</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">20</span>, height<span class="op">=</span><span class="fl">20</span><span class="op">*</span><span class="op">(</span><span class="fl">2</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span>, res<span class="op">=</span><span class="fl">400</span>, units<span class="op">=</span><span class="st">'in'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><p><img src="figures/projection/visium_preservation_all.png" width="700" height="700"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
