<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>ConstructNetwork — ConstructNetwork • hdWGCNA</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="ConstructNetwork — ConstructNetwork"><meta property="og:description" content="Constructs a co-expression network and groups genes into modules
given a Seurat object that has been prepared for network analysis."><meta property="og:image" content="https://smorabit.github.io/hdWGCNA/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hdWGCNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/hdWGCNA.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Core functionality</li>
    <li>
      <a href="../articles/basic_tutorial.html">hdWGCNA in single-cell data</a>
    </li>
    <li>
      <a href="../articles/ST_basics.html">hdWGCNA in spatial transcriptomics data</a>
    </li>
    <li>
      <a href="../articles/network_visualizations.html">Network visualization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Biological context for co-expression modules</li>
    <li>
      <a href="../articles/differential_MEs.html">Differential module eigengene (DME) analysis</a>
    </li>
    <li>
      <a href="../articles/module_trait_correlation.html">Module trait correlation</a>
    </li>
    <li>
      <a href="../articles/enrichment_analysis.html">Enrichment analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Exploring modules in external datasets</li>
    <li>
      <a href="../articles/projecting_modules.html">Projecting modules to new datasets</a>
    </li>
    <li>
      <a href="../articles/module_preservation.html">Module preservation and reproducibility</a>
    </li>
    <li>
      <a href="../articles/projecting_modules_cross.html">Cross-species and cross-modality analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced topics</li>
    <li>
      <a href="../articles/consensus_wgcna.html">Consensus network analysis</a>
    </li>
    <li>
      <a href="../articles/motif_analysis.html">Motif analysis</a>
    </li>
    <li>
      <a href="../articles/other_metacells.html">Alternative metacell algorithms</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/customization.html">Module customization</a>
    </li>
    <li>
      <a href="../articles/sctransform.html">Using SCTransform normalized data</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/index.html">All vignettes</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>ConstructNetwork</h1>
    
    <div class="hidden name"><code>ConstructNetwork.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Constructs a co-expression network and groups genes into modules
given a Seurat object that has been prepared for network analysis.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">ConstructNetwork</span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  soft_power <span class="op">=</span> <span class="cn">NULL</span>,
  min_power <span class="op">=</span> <span class="fl">3</span>,
  tom_outdir <span class="op">=</span> <span class="st">"TOM"</span>,
  tom_name <span class="op">=</span> <span class="cn">NULL</span>,
  consensus <span class="op">=</span> <span class="cn">FALSE</span>,
  overwrite_tom <span class="op">=</span> <span class="cn">FALSE</span>,
  wgcna_name <span class="op">=</span> <span class="cn">NULL</span>,
  blocks <span class="op">=</span> <span class="cn">NULL</span>,
  maxBlockSize <span class="op">=</span> <span class="fl">30000</span>,
  randomSeed <span class="op">=</span> <span class="fl">12345</span>,
  corType <span class="op">=</span> <span class="st">"pearson"</span>,
  consensusQuantile <span class="op">=</span> <span class="fl">0.3</span>,
  networkType <span class="op">=</span> <span class="st">"signed"</span>,
  TOMType <span class="op">=</span> <span class="st">"signed"</span>,
  TOMDenom <span class="op">=</span> <span class="st">"min"</span>,
  scaleTOMs <span class="op">=</span> <span class="cn">TRUE</span>,
  scaleQuantile <span class="op">=</span> <span class="fl">0.8</span>,
  sampleForScaling <span class="op">=</span> <span class="cn">TRUE</span>,
  sampleForScalingFactor <span class="op">=</span> <span class="fl">1000</span>,
  useDiskCache <span class="op">=</span> <span class="cn">TRUE</span>,
  chunkSize <span class="op">=</span> <span class="cn">NULL</span>,
  deepSplit <span class="op">=</span> <span class="fl">4</span>,
  pamStage <span class="op">=</span> <span class="cn">FALSE</span>,
  detectCutHeight <span class="op">=</span> <span class="fl">0.995</span>,
  minModuleSize <span class="op">=</span> <span class="fl">50</span>,
  mergeCutHeight <span class="op">=</span> <span class="fl">0.2</span>,
  saveConsensusTOMs <span class="op">=</span> <span class="cn">TRUE</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>seurat_obj</dt>
<dd><p>A Seurat object</p></dd>
<dt>soft_power</dt>
<dd><p>the soft power used for network construction. Automatically selected by default.</p></dd>
<dt>min_power</dt>
<dd><p>the smallest soft power to be selected if soft_power=NULL</p></dd>
<dt>tom_outdir</dt>
<dd><p>path to the directory where the TOM will be written</p></dd>
<dt>tom_name</dt>
<dd><p>prefix name given to the TOM output file</p></dd>
<dt>consensus</dt>
<dd><p>flag indicating whether or not to perform Consensus network analysis</p></dd>
<dt>wgcna_name</dt>
<dd><p>name of the WGCNA experiment</p></dd>
<dt>...</dt>
<dd><p>additional parameters passed to blockwiseConsensusModules</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>seurat_obj with the co-expression network and gene modules computed for the selected wgcna experiment</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>ConstructNetwork builds a co-expression network and identifies clusters of
highly co-expressed genes (modules) from the metacell or metaspot
expression matrix stored in the Seurat object. Before running this function,
the following functions must be run on the input Seurat object:</p><ol><li><p>SetupForWGCNA</p></li>
<li><p>MetacellsByGroups or MetaspotsByGroups, and NormalizeMetacells</p></li>
<li><p>SetDatExpr or SetMultiExpr</p></li>
<li><p>TestSoftPowers or TestSoftPowersConsensus</p></li>
</ol><p>This function can also be used to perform consensus network analysis if consensus=TRUE
is selected. ConstructNetwork calls the WGCNA function blockwiseConsensusModules
to compute the adjacency matrix, topological overlap matrix, and to run the
Dynamic Tree Cut algorithm to identify gene modules. blockwiseConsensusModules
has numerous parameters but here we have selected default parameters that we
have found to provide reasonable results on a variety of single-cell and
spatial transcriptomic datasets.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

