<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>OverlapModulesDEGs — OverlapModulesDEGs • hdWGCNA</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="OverlapModulesDEGs — OverlapModulesDEGs"><meta property="og:description" content="Performs Fisher's Exact Test for overlap between DEGs and hdWGCNA modules."><meta property="og:image" content="https://smorabit.github.io/hdWGCNA/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hdWGCNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/hdWGCNA.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Core functionality</li>
    <li>
      <a href="../articles/basic_tutorial.html">hdWGCNA in single-cell data</a>
    </li>
    <li>
      <a href="../articles/ST_basics.html">hdWGCNA in spatial transcriptomics data</a>
    </li>
    <li>
      <a href="../articles/isoform_pbmcs.html">Isoform co-expression network analysis with PacBio MAS-Seq</a>
    </li>
    <li>
      <a href="../articles/network_visualizations.html">Network visualization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Biological context for co-expression modules</li>
    <li>
      <a href="../articles/differential_MEs.html">Differential module eigengene (DME) analysis</a>
    </li>
    <li>
      <a href="../articles/module_trait_correlation.html">Module trait correlation</a>
    </li>
    <li>
      <a href="../articles/enrichment_analysis.html">Enrichment analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Exploring modules in external datasets</li>
    <li>
      <a href="../articles/projecting_modules.html">Projecting modules to new datasets</a>
    </li>
    <li>
      <a href="../articles/module_preservation.html">Module preservation and reproducibility</a>
    </li>
    <li>
      <a href="../articles/projecting_modules_cross.html">Cross-species and cross-modality analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced topics</li>
    <li>
      <a href="../articles/consensus_wgcna.html">Consensus network analysis</a>
    </li>
    <li>
      <a href="../articles/pseudotime.html">Co-expression module dynamics with pseudotime</a>
    </li>
    <li>
      <a href="../articles/motif_analysis.html">Motif analysis</a>
    </li>
    <li>
      <a href="../articles/other_metacells.html">Alternative metacell algorithms</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/customization.html">Module customization</a>
    </li>
    <li>
      <a href="../articles/sctransform.html">Using SCTransform normalized data</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/index.html">All vignettes</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>OverlapModulesDEGs</h1>
    
    <div class="hidden name"><code>OverlapModulesDEGs.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Performs Fisher's Exact Test for overlap between DEGs and hdWGCNA modules.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">OverlapModulesDEGs</span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  <span class="va">deg_df</span>,
  fc_cutoff <span class="op">=</span> <span class="fl">0.5</span>,
  group_col <span class="op">=</span> <span class="st">"cluster"</span>,
  wgcna_name <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>seurat_obj</dt>
<dd><p>A Seurat object</p></dd>
<dt>deg_df</dt>
<dd><p>DEG table formatted like the output from Seurat's FindMarkers</p></dd>
<dt>fc_cutoff</dt>
<dd><p>log fold change cutoff for DEGs to be included in the overlap test</p></dd>
<dt>group_col</dt>
<dd><p>the name of the column in deg_df containing the cell grouping information</p></dd>
<dt>wgcna_name</dt>
<dd><p>The name of the hdWGCNA experiment in the seurat_obj@misc slot</p></dd>
</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">OverlapModulesDEGs</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> function (seurat_obj, deg_df, fc_cutoff = 0.5, group_col = "cluster", </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     wgcna_name = NULL, ...) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (is.null(wgcna_name)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         wgcna_name &lt;- seurat_obj@misc$active_wgcna</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     deg_df$group &lt;- deg_df[, group_col]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     cell_groups &lt;- deg_df$group %&gt;% unique</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     modules &lt;- GetModules(seurat_obj)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mods &lt;- levels(modules$module)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mods &lt;- mods[mods != "grey"]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (fc_cutoff &gt;= 0) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         deg_df &lt;- subset(deg_df, avg_log2FC &gt;= fc_cutoff)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     else {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         deg_df &lt;- subset(deg_df, avg_log2FC &lt;= fc_cutoff)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         deg_df$avg_log2FC &lt;- -1 * deg_df$avg_log2FC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         fc_cutoff &lt;- -1 * fc_cutoff</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     genome.size &lt;- nrow(seurat_obj)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     overlap_df &lt;- do.call(rbind, lapply(mods, function(cur_mod) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         cur_module_genes &lt;- modules %&gt;% subset(module == cur_mod) %&gt;% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             .$gene_name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         cur_overlap_df &lt;- do.call(rbind, lapply(cell_groups, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             function(cur_group) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 cur_DEGs &lt;- deg_df %&gt;% subset(group == cur_group &amp; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   p_val_adj &lt;= 0.05 &amp; avg_log2FC &gt; fc_cutoff) %&gt;% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   .$gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 cur_overlap &lt;- testGeneOverlap(newGeneOverlap(cur_module_genes, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   cur_DEGs, genome.size = genome.size))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 c(cur_overlap@odds.ratio, cur_overlap@pval, cur_overlap@Jaccard, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   length(cur_overlap@intersection))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             })) %&gt;% as.data.frame</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         colnames(cur_overlap_df) &lt;- c("odds_ratio", "pval", "Jaccard", </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             "size_intersection")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         cur_overlap_df$module &lt;- cur_mod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         cur_overlap_df$group &lt;- cell_groups</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         cur_overlap_df$color &lt;- modules %&gt;% subset(module == </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             cur_mod) %&gt;% .$color %&gt;% unique</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         cur_overlap_df</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     overlap_df$fdr &lt;- p.adjust(overlap_df$pval, method = "fdr")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     overlap_df$Significance &lt;- gtools::stars.pval(overlap_df$fdr)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     overlap_df$Significance &lt;- ifelse(overlap_df$Significance == </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         ".", "", overlap_df$Significance)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     overlap_df$module &lt;- factor(overlap_df$module, levels = mods)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     overlap_df &lt;- overlap_df %&gt;% dplyr::select(c(module, group, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         color, odds_ratio, pval, fdr, Significance, Jaccard, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         size_intersection))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     overlap_df</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;bytecode: 0x7fd6e118daa0&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;environment: namespace:hdWGCNA&gt;</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

