<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>PlotModulePreservation — PlotModulePreservation • hdWGCNA</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="PlotModulePreservation — PlotModulePreservation"><meta property="og:description" content="Plotting function for Module Preservation statistics"><meta property="og:image" content="https://smorabit.github.io/hdWGCNA/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hdWGCNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.02</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/hdWGCNA.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Core functionality</li>
    <li>
      <a href="../articles/basic_tutorial.html">hdWGCNA in single-cell data</a>
    </li>
    <li>
      <a href="../articles/ST_basics.html">hdWGCNA in spatial transcriptomics data</a>
    </li>
    <li>
      <a href="../articles/network_visualizations.html">Network visualization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Biological context for co-expression modules</li>
    <li>
      <a href="../articles/differential_MEs.html">Differential module eigengene (DME) analysis</a>
    </li>
    <li>
      <a href="../articles/module_trait_correlation.html">Module trait correlation</a>
    </li>
    <li>
      <a href="../articles/enrichment_analysis.html">Enrichment analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Exploring modules in external datasets</li>
    <li>
      <a href="../articles/projecting_modules.html">Projecting modules to new datasets</a>
    </li>
    <li>
      <a href="../articles/module_preservation.html">Module preservation and reproducibility</a>
    </li>
    <li>
      <a href="../articles/projecting_modules_cross.html">Cross-species and cross-modality analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced topics</li>
    <li>
      <a href="../articles/consensus_wgcna.html">Consensus network analysis</a>
    </li>
    <li>
      <a href="../articles/motif_analysis.html">Motif analysis</a>
    </li>
    <li>
      <a href="../articles/other_metacells.html">Alternative metacell algorithms</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/customization.html">Module customization</a>
    </li>
    <li>
      <a href="../articles/sctransform.html">Using SCTransform normalized data</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/index.html">All vignettes</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>PlotModulePreservation</h1>
    
    <div class="hidden name"><code>PlotModulePreservation.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plotting function for Module Preservation statistics</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">PlotModulePreservation</span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  <span class="va">name</span>,
  statistics <span class="op">=</span> <span class="st">"summary"</span>,
  plot_labels <span class="op">=</span> <span class="cn">TRUE</span>,
  label_size <span class="op">=</span> <span class="fl">4</span>,
  mod_point_size <span class="op">=</span> <span class="fl">4</span>,
  wgcna_name <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>seurat_obj</dt>
<dd><p>A Seurat object</p></dd>
<dt>name</dt>
<dd><p>The name of the module preservation analysis to plot given in ModulePreservation</p></dd>
<dt>statistics</dt>
<dd><p>Which module preservation statistics to plot? Choices are summary, all, or a custom list</p></dd>
<dt>plot_labels</dt>
<dd><p>logical determining whether to plot the module labels</p></dd>
<dt>label_size</dt>
<dd><p>the size of the module labels</p></dd>
<dt>mod_point_size</dt>
<dd><p>the size of the points in each plot</p></dd>
<dt>wgcna_name</dt>
<dd><p>The name of the hdWGCNA experiment in the seurat_obj@misc slot</p></dd>
</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">PlotModulePreservation</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> function (seurat_obj, name, statistics = "summary", plot_labels = TRUE, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     label_size = 4, mod_point_size = 4, wgcna_name = NULL) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (is.null(wgcna_name)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         wgcna_name &lt;- seurat_obj@misc$active_wgcna</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mod_pres &lt;- GetModulePreservation(seurat_obj, name, wgcna_name)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     obs_df &lt;- mod_pres$obs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Z_df &lt;- mod_pres$Z</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     modules &lt;- GetModules(seurat_obj, wgcna_name)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     module_colors &lt;- modules %&gt;% dplyr::select(c(module, color)) %&gt;% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         distinct</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mods &lt;- rownames(Z_df)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mod_colors &lt;- module_colors$color[match(mods, module_colors$module)]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mod_colors = ifelse(is.na(mod_colors), "gold", mod_colors)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (statistics == "summary") {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         stat_list &lt;- c("Zsummary.qual", "Zsummary.pres")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     else if (statistics == "rank") {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         stat_list &lt;- colnames(obs_df[, -1])[grepl("Rank", colnames(obs_df[, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             -1]))]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     else if (statistics == "all") {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         stat_list &lt;- c(colnames(obs_df[, -1])[grepl("Rank", colnames(obs_df[, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             -1]))], colnames(Z_df[, -1]))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     else {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         stat_list &lt;- statistics</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     stat_list &lt;- stat_list[stat_list != "moduleSize"]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     plot_list &lt;- list()</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     for (statistic in stat_list) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         print(statistic)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         if (statistic %in% colnames(obs_df)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             values &lt;- obs_df[, statistic]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         else if (statistic %in% colnames(Z_df)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             values &lt;- Z_df[, statistic]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         else {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             stop("Invalid name for statistic.")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         plot_df &lt;- data.frame(module = mods, color = mod_colors, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             value = values, size = Z_df$moduleSize)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         plot_df &lt;- plot_df %&gt;% subset(!(module %in% c("grey", </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             "gold")))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         if (grepl("Rank", statistic)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             cur_p &lt;- plot_df %&gt;% ggplot(aes(x = size, y = value, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 fill = module, color = module)) + geom_point(size = mod_point_size, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 pch = 21, color = "black") + scale_y_reverse()</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         else {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             cur_p &lt;- plot_df %&gt;% ggplot(aes(x = size, y = value, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 fill = module, color = module)) + geom_rect(data = plot_df[1, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 ], aes(xmin = 0, xmax = Inf, ymin = -Inf, ymax = 2), </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 fill = "grey75", alpha = 0.8, color = NA) + geom_rect(data = plot_df[1, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 ], aes(xmin = 0, xmax = Inf, ymin = 2, ymax = 10), </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 fill = "grey92", alpha = 0.8, color = NA) + geom_point(size = mod_point_size, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 pch = 21, color = "black")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         cur_p &lt;- cur_p + scale_fill_manual(values = plot_df$color) + </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             scale_color_manual(values = plot_df$color) + scale_x_continuous(trans = "log10") + </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             ylab(statistic) + xlab("Module Size") + ggtitle(statistic) + </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             NoLegend() + theme(plot.title = element_text(hjust = 0.5))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         if (plot_labels) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             cur_p &lt;- cur_p + geom_text_repel(label = plot_df$module, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 size = label_size, max.overlaps = Inf, color = "black")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         plot_list[[statistic]] &lt;- cur_p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (length(plot_list) == 1) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         return(plot_list[[1]])</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     plot_list</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;bytecode: 0x7fe32ae7ba30&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;environment: namespace:hdWGCNA&gt;</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

