<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>ProjectModules — ProjectModules • hdWGCNA</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="ProjectModules — ProjectModules"><meta property="og:description" content="Project a set of co-expression modules from a reference to a query dataset"><meta property="og:image" content="https://smorabit.github.io/hdWGCNA/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hdWGCNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/hdWGCNA.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Core functionality</li>
    <li>
      <a href="../articles/basic_tutorial.html">hdWGCNA in single-cell data</a>
    </li>
    <li>
      <a href="../articles/ST_basics.html">hdWGCNA in spatial transcriptomics data</a>
    </li>
    <li>
      <a href="../articles/isoform_pbmcs.html">Isoform co-expression network analysis with PacBio MAS-Seq</a>
    </li>
    <li>
      <a href="../articles/network_visualizations.html">Network visualization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Biological context for co-expression modules</li>
    <li>
      <a href="../articles/differential_MEs.html">Differential module eigengene (DME) analysis</a>
    </li>
    <li>
      <a href="../articles/module_trait_correlation.html">Module trait correlation</a>
    </li>
    <li>
      <a href="../articles/enrichment_analysis.html">Enrichment analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Exploring modules in external datasets</li>
    <li>
      <a href="../articles/projecting_modules.html">Projecting modules to new datasets</a>
    </li>
    <li>
      <a href="../articles/module_preservation.html">Module preservation and reproducibility</a>
    </li>
    <li>
      <a href="../articles/projecting_modules_cross.html">Cross-species and cross-modality analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced topics</li>
    <li>
      <a href="../articles/consensus_wgcna.html">Consensus network analysis</a>
    </li>
    <li>
      <a href="../articles/pseudotime.html">Co-expression module dynamics with pseudotime</a>
    </li>
    <li>
      <a href="../articles/motif_analysis.html">Motif analysis</a>
    </li>
    <li>
      <a href="../articles/other_metacells.html">Alternative metacell algorithms</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/customization.html">Module customization</a>
    </li>
    <li>
      <a href="../articles/sctransform.html">Using SCTransform normalized data</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/index.html">All vignettes</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>ProjectModules</h1>
    
    <div class="hidden name"><code>ProjectModules.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Project a set of co-expression modules from a reference to a query dataset</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">ProjectModules</span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  <span class="va">seurat_ref</span>,
  modules <span class="op">=</span> <span class="cn">NULL</span>,
  group.by.vars <span class="op">=</span> <span class="cn">NULL</span>,
  gene_mapping <span class="op">=</span> <span class="cn">NULL</span>,
  genome1_col <span class="op">=</span> <span class="cn">NULL</span>,
  genome2_col <span class="op">=</span> <span class="cn">NULL</span>,
  overlap_proportion <span class="op">=</span> <span class="fl">0.5</span>,
  vars.to.regress <span class="op">=</span> <span class="cn">NULL</span>,
  scale.model.use <span class="op">=</span> <span class="st">"linear"</span>,
  wgcna_name <span class="op">=</span> <span class="cn">NULL</span>,
  wgcna_name_proj <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>seurat_obj</dt>
<dd><p>A Seurat object where the modules will be projected to</p></dd>
<dt>seurat_ref</dt>
<dd><p>A Seurat object containing the co-expression modules to be projected</p></dd>
<dt>modules</dt>
<dd><p>optionally provide a dataframe containing gene-module information to bypass seurat_ref</p></dd>
<dt>group.by.vars</dt>
<dd><p>groups to harmonize by</p></dd>
<dt>gene_mapping</dt>
<dd><p>dataframe to map gene names between genomes</p></dd>
<dt>genome1_col</dt>
<dd><p>column in gene_mapping containing the genes for seurat_ref (reference)</p></dd>
<dt>genome2_col</dt>
<dd><p>column in gene_mapping containing the genes for seurat_obj (query)</p></dd>
<dt>overlap_proportion</dt>
<dd><p>the proportion of genes that must be present in seurat_obj for a given module to be projected. Default = 0.5 (50% of genes)</p></dd>
<dt>vars.to.regress</dt>
<dd><p>character vector of variables in seurat_obj@meta.data to regress when running ScaleData</p></dd>
<dt>scale.model.use</dt>
<dd><p>model to scale data when running ScaleData choices are "linear", "poisson", or "negbinom"</p></dd>
<dt>wgcna_name</dt>
<dd><p>The name of the hdWGCNA experiment in the seurat_ref@misc slot</p></dd>
<dt>wgcna_name_proj</dt>
<dd><p>The name of the hdWGCNA experiment to be created for the projected modules in seurat_obj</p></dd>
</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">ProjectModules</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> function (seurat_obj, seurat_ref, modules = NULL, group.by.vars = NULL, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     gene_mapping = NULL, genome1_col = NULL, genome2_col = NULL, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     overlap_proportion = 0.5, vars.to.regress = NULL, scale.model.use = "linear", </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     wgcna_name = NULL, wgcna_name_proj = NULL, ...) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (is.null(wgcna_name)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         wgcna_name &lt;- seurat_ref@misc$active_wgcna</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (is.null(modules)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         modules &lt;- GetModules(seurat_ref, wgcna_name)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     else {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         if (!all(c("gene_name", "module", "color") %in% colnames(modules))) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             stop("Missing columns in modules table. Required columns are gene_name, module, color")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (!is.factor(modules$module)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         modules$module &lt;- as.factor(modules$module)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (!is.null(gene_mapping)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         modules &lt;- TransferModuleGenome(modules, gene_mapping, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             genome1_col, genome2_col)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mods &lt;- as.character(unique(modules$module))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mod_props &lt;- unlist(lapply(mods, function(x) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         cur_mod &lt;- subset(modules, module == x)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         sum(cur_mod$gene_name %in% rownames(seurat_obj))/nrow(cur_mod)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mods_keep &lt;- mods[mod_props &gt;= overlap_proportion]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mods_remove &lt;- mods[mod_props &lt; overlap_proportion]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (length(mods) &gt; 0) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         warning(paste0("The following modules will not be projected because too few genes are present in seurat_obj: ", </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             paste(mods_remove, collapse = ", ")))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     modules &lt;- subset(modules, module %in% mods_keep) %&gt;% dplyr::mutate(module = droplevels(module))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     gene_names &lt;- modules$gene_name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     genes_use &lt;- intersect(gene_names, rownames(seurat_obj))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     modules &lt;- modules %&gt;% subset(gene_name %in% genes_use)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (!(wgcna_name_proj %in% names(seurat_obj@misc))) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         seurat_obj &lt;- SetupForWGCNA(seurat_obj, wgcna_name = wgcna_name_proj, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             features = genes_use)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     print("project modules")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     seurat_obj &lt;- ModuleEigengenes(seurat_obj, group.by.vars = group.by.vars, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         modules = modules, vars.to.regress = vars.to.regress, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         scale.model.use = scale.model.use, wgcna_name = wgcna_name_proj, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         ...)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     seurat_obj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;bytecode: 0x7ffe78405188&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;environment: namespace:hdWGCNA&gt;</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

